% Glossaries are implemented using the `glossaries' package.

% How to define homographs - see details in glossaries-manual.html distributed with the 'glossaries' package.
%\newglossaryentry{glossary}{name=glossary,  description={\nopostdesc},  plural={glossaries}
%\newglossaryentry{glossarylist}{ description={1) list of technical words},  sort={1},  parent={glossary}}  
%\newglossaryentry{glossarycol}{ description={2) collection of glosses},  sort={2},  parent={glossary}}

\newglossaryentry{abm}{name={Agent-Based Model},description={See \gls{ibm}}}
\newglossaryentry{achem}{name={Artificial Chemistry},text={artificial chemistry}, plural={artificial chemistries},description={The computational abstraction of biomolecular processes}}
\newacronym{acs}{ACS}{Autocatalytic Set}
\newacronym{ae}{AE}{Artificial Evolution}
\newacronym{ais}{AIS}{Artificial Immune System}
\newacronym{alife}{Alife}{Artificial Life}
\newacronym{ann}{ANN}{Artificial Neural Network}
\newacronym{anova}{ANOVA}{Analysis of Variance}
\newacronym{aoc}{AOC}{Arrow of Complexity}
\newglossaryentry{arthropod}{name={Arthropod},text={arthropod},description={An invertebrate animal with segmented body, exoskeleton, and jointed appendages. Examples include insects and crustaceans}}

\newglossaryentry{baldwin}{name={Baldwin effect},description={Learning can provide an easy evolutionary path towards co-adapted alleles in environments that have no good evolutionary path for non-learning organisms \parencite{Hinton:1987vy}}}
\newacronym{bn}{BN}{Boolean Network}

\newglossaryentry{canalisation}{name={Canalisation},text={canalisation},description={``...developmental reactions, as they occur in organisms submitted to natural selection...are adjusted so as to bring about one definite end-result regardless of minor variations in conditions during the course of the reaction.'' \parencite{Waddington:1942jb}}}
\newglossaryentry{constructive}{name={Constructive}, text={constructive}, description={An artificial chemistry is constructive if it is closed under all reactions. That is, if the products of a reaction may become reactants in turn.}}
\newglossaryentry{crm}{name={CRM (cis-regulatory module)},description={Typically relatively short stretches of DNA (~ 300-500 bp), which integrate multiple inputs from different TFs to give rise to a distinct output of spatio-temporal expression \parencite{Wilczynski:2010rt}}}
\newacronym{ca}{CA}{Cellular Automaton}
\newacronym{ci}{CI}{Computational Intelligence}
\newglossaryentry{constitutively}{name={constitutively},description={A gene that is transcribed continuously; a constant process}}

\newacronym{dbn}{DBN}{Dynamic Bayesian Network}
\newglossaryentry{deephomology}{name={deep homology},description={Morphologically disparate organs whose formation (and evolution) depends on homologous genetic regulatory circuits \parencite{Shubin:2009vw}}}
\newglossaryentry{designVariables}{name={design variables},description={Vector of parameters that can be changed by an algorithm}}
\newacronym{dfa}{DFA}{Detrended Fluctuation Analysis}
\newglossaryentry{drosophila}{name={Drosophila},description={Genus Drosophila, fruit flies or small flies. The species \emph{Drosophila melanogaster}, the common fruit fly, is a standard model organism in biology and has been extensively studied}}

\newacronym{ea}{EA}{Evolutionary Algorithm}
\newacronym{ens}{ENS}{Evolution by Natural Selection}
\newglossaryentry{environmentalVariables}{name={environmental variables},description={All parameters to an algorithm that are not design variables; those parameters that are not in the control of an algorithm but none-the-less influence its operation}}
\newglossaryentry{epistasis}{name={epistasis},description={The interaction between two or more genes at different loci such that phenotypic expression of one depends on expression of another. \parencite{OUPBioMolBio:fr}}}
\newglossaryentry{epigenetic}{name={epigenetic}, description={\nopostdesc}}
\newglossaryentry{epigenetic1}{description={1) ``describing any of the mechanisms regulating the expression and interaction of genes, particularly during the development process. These include changes that influence the phenotype but have arisen as a result of mechanisms such as inherited patterns of DNA methylation rather than differences in gene sequence: imprinting is an example of this.'' \parencite{OUPBioMolBio:fr}},sort={1}, parent={epigenetic}}
\newglossaryentry{epigenetic2}{description={2) ``describing heritable changes that are not the result of changes in DNA sequence.'' \parencite{OUPBiol:mz}},sort={2},parent={epigenetic}}
\newglossaryentry{eukaryote}{name={eukaryote},description={An organism consisting of cells in which the genetic material is contained within a distinct nucleus. \parencite{OUPBiol:mz}}}
\newacronym{evoevo}{EvoEvo}{Evolution-of-evolution}
\newglossaryentry{evolvability}{name={evolvability},description={\nopostdesc}}
\newglossaryentry{evolvability1}{description={1) ``A biological system is evolvable if it can acquire novel functions (phenotypes) through genetic change (pertubations), functions that help the organism survive and reproduce.'' \parencite{Wagner:2005jw}},sort={1}, parent={evolvability}}
\newglossaryentry {evolvability2}{description={2) ``the efficiency of an organism in discovering beneficial mutants (the rate or degree of evolution)'' (\cite{Crombach:2008br})},sort={1}, parent={evolvability}}
\newglossaryentry {evolvability3}{description={3) ``the system is evolvable if mutations in it can produce heritable phenotypic variation.'' (\cite{Wagner:2008mi})},sort={2}, parent={evolvability}}
\newglossaryentry {evolvability4}{description={4) ``the genome's ability to produce adaptive variants when acted upon by the genetic system.'' (\cite{Wagner:1996kc})},sort={3}, parent={evolvability}}
\newglossaryentry {evolvability5}{description={5) ``...the network is evolvable if, as a result of these mutations [the duplication and divergence of one or more genes] new attractors emerge.'' \parencite{Aldana:2007da}},sort={4}, parent={evolvability}}
\newglossaryentry{experiment}{name={experiment},description={``\ldots a test or series of tests in which purposeful changes are made to the input variables of a process or system so that we may observe and identify the reasons for changes that may be observed in the output response.'' \parencite{Montgomery2009}}}

\newglossaryentry{foodset}{name={food set},description={A set of molecules maintained at a constant concentration that acts as an always-available source of reactants (see in \cite[p313]{Kauffman1993} (original reference) and \cite{Hordijk2004}}}

\newglossaryentry{genotype}{name={Genotype},text={genotype},description={The heritable definition of an entity, that when mapped through a G$\rightarrow$P map produces a phenotype}}
\newglossaryentry{gpmap}{name={G$\rightarrow$P map},description={Genotype$\rightarrow$phenotype map or the relationship Genotype-Phenotype: ``In the theoretical scheme proposed by evolutionary genetics, development is the function that maps the genotype onto the phenotype.'' \parencite{Alberch:1991ve}. `Map` is used exactly in the mathematical sense}}
\newacronym{grn}{GRN}{Gene Regulatory Network}

\newglossaryentry{heredity}{name={Heredity},text={heredity}, description={The passing on of traits from parent to offspring}}
\newglossaryentry{heterochrony}{name={heterochrony},description={A change in the timing of a developmental event (\eg time of appearance of a morphological trait or of expression of a gene) relative to other developmental event(s) \parencite{OUPGen:qf}}}
\newglossaryentry{heritability}{name={Heritability}, text={heritability}, description={The proportion of the total variation between individuals in a given population due to genetic variation, or the degree to which environment or genetics influences the phenotype. (see \url{https://www.ncbi.nlm.nih.gov/books/NBK22001/})}}
\newglossaryentry{homeostasis}{name={homeostasis},description={Coined by Walter B. Cannon \parencite{Cannon:1926zr}.The tendency of a biological system to resist change and to maintain itself in a state of stable equilibrium \parencite{OUPZoo:ys}}}
\newglossaryentry{homology}{name={homology},description={\nopostdesc}}
\newglossaryentry{homology1}{description={1) historical continuity in which morphological features in related species are similar in pattern or form because they evolved from a corresponding structure in a common ancestor. \parencite{Shubin:2009vw}},sort={1}, parent={homology}}
\newglossaryentry{homology2}{description={2) Deep homology: ``the sharing of the genetic regulatory apparatus that is used to build morphologically and phylogenetically disparate animal features \parencite{Shubin:2009vw}},sort={2}, parent={homology}}
\newglossaryentry{hgt}{name={Horizontal Gene Transfer},description={``The transfer of genetic material between reproductively isolated species'' \parencite{Pace:2008vi}}}
\newglossaryentry{hypercycle}{name={Hypercycle},text={hypercycle}, description={An organisation of self-replicating molecules connected in an autocatalytic cycle \parencite{Eigen1971}.}}

\newglossaryentry{ibm}{name={Individual-Based Model},description={\quote{In principle, IBMs simulate populations or systems of populations as being composed of discrete agents that represent individual organisms or groups of similar individual organisms, with sets of traits that vary among the agents. Each agent has a unique history of interactions with its environment and other agents. IBMs attempt to capture the variation among individuals that is relevant to the questions being addressed.}{\parencite{DeAngelis2005}}}}
\newacronym{ida}{IDA}{Initial Darwinian Ancestor}
\newacronym{ke}{KE}{Kinetic Energy}

\newglossaryentry{lamarck}{name={Lamarckian evolution},description={The heritability of acquired characteristics}}
\newglossaryentry{landscape}{name={fitness landscape}, description={A relationship between a set of genes (or a set of quantitative characters) and a measure of fitness (e.g. viability, fertility, or mating success) \parencite{Gravner:2007yd}}}
\newacronym{luca}{LUCA}{Last Universal Common Ancestor}

\newglossaryentry{machinelearning}{name={machine learning},description={``A computer program is said to learn from experience E with respect to some class of tasks T and performance measure P, if its performance at tasks in T, as measured by P, improves with experience E.'' \parencite[p.2]{Mitchell:1997fk}}}
\newglossaryentry{metathorax}{name={Metathorax},text={metathorax},description={The posterior-most segment of the thorax. Bears the hindwings in most insects, and the third pair of legs}}
\newacronym{moea}{MOEA}{Multiple-Objective Evolutionary Algorithm}
\newglossaryentry{monotonal}{name={monotonal},description={Where the sign of $f'(x)$ is constant}}
\newglossaryentry{morphogenesis}{name={morphogenesis},description={The development of the body plan or morphology}}

\newglossaryentry{naturalselection}{name={Natural Selection},text={natural selection},description={The differential survival and reproduction of genotypes \parencite[pg. 26]{Futuyama:1979tg}}}
\newglossaryentry{neutralClusters}{name={neutral clusters},description={Connected components of same fitness \parencite{Gravner:2007yd}}}

\newglossaryentry{oe}{name={Open-Ended Evolution},text={open-ended evolution},description={``the ability to continuously produce novelty and/or complexity'' \parencite{BanzhafBaumgaertnerBeslonEtAl2016}}}
\newglossaryentry{ontogeny}{name={ontogeny},description={``The sequence of events in the development of an individual organism during its lifetime.'' \parencite{OUPBioMolBio:fr}}}
\newacronym{ode}{ODE}{Ordinary Differential Equation}	

\newglossaryentry{phenotype}{name={Phenotype},text={phenotype},description={``The observable manifestation of a specific genotype; those properties of an organism produced by the genotype in conjunction with the environment that are observable. Organisms with the same overall genotype may have different phenotypes because of the effects of the environment and of gene interaction. Conversely, organisms may have the same phenotype but different genotypes, as a result of incomplete dominance, penetrance, or expressivity.'' \parencite{OUPZoo:ys}}}
\newglossaryentry{plasticity}{name={phenotypic plasticity},description={\nopostdesc}}
\newglossaryentry{plasticity1}{description={1) ``The ability of individual genotypes to produce different phenotypes when exposed to different environmental conditions'' \parencite{}},sort={1},parent={plasticity}}
\newglossaryentry{plasticity2}{description={2) ``The ability of an organism to adapt to changes in its environment by modifying its own development, form, behaviour, or other trait.'' \parencite{OUPBiol:mz}},sort={2},parent={plasticity}}
\newglossaryentry{plasticity3}{description={3) ``A general term that covers all types of environmentally induced phenotypic variation.'' (\cite{Stearns:1989gf})},sort={3},parent={plasticity}}
 \newglossaryentry{plasticity4}{description={4) ``The degree of change in phenotype against variation in environment.'' \parencite{Kaneko:2009jx}},sort={4},parent={plasticity}}
\newglossaryentry{pleiotropy}{name={pleiotropy},description={``An allele that has more than one effect in an organism.'' \parencite{OUPBiol:mz}}}
\newglossaryentry{polyphenism}{name={Polyphenism},text={polyphenism},description={Alternative phenotypes produced under the effect of environmental cues \parencite{Minelli:2010vn}}}
\newglossaryentry{prokaryote}{name={prokaryote},description={``Any organism in which the genetic material is not enclosed in a cell nucleus.'' \parencite{OUPBiol:mz}}}
\newacronym{prng}{PRNG}{Pseudo-Random Number Generator}
\newacronym{qq}{Q-Q}{quantile-quantile}

\newacronym{raf}{RAF}{Reflexively autocatalytic and F-generated}
\newglossaryentry{replicate}{name={Replicate},text={replicate},plural={replicates},description={``\ldots an independent repeat of each factor combination.'' \parencite{Montgomery2009}}}
\newglossaryentry{replicon}{name={replicon},description={``A genetic element that behaves as an autonomous unit during DNA replication.'' \parencite{OUPGen:qf}}}
\newacronym{rbn}{RBN}{Random Boolean Network}
\newglossaryentry{rdkit}{name={RDKit},description={Open-source software package for chemical kinematics, available from \url{http://www.rdkit.org}}}
\newacronym{rnn}{RNN}{Recursive Neural Network}
\newglossaryentry{robustness}{name={robustness},description={\nopostdesc}}
\newglossaryentry{robustness1}{description={1) ``A property that allows a system to maintain its functions despite external and internal perturbations.'' \parencite{Kitano:2004yy}},sort={1}, parent={robustness}}
\newglossaryentry{robustness2}{description={2) ``A solution should still work satisfactorily when the design variables change slightly [after the optimal solution has been determined]'' \parencite{Jin:2005zr}},sort={2}, parent={robustness}}
\newglossaryentry{robustness3}{description={3) ``A biological system [is] mutationally robust if its function or structure persists after mutations in its parts.'' \parencite{Wagner:2008mi}},sort={3}, parent={robustness}}
\newglossaryentry{robustness4}{description={4) ``The invariance of phenotypes in the face of perturbation'' where pertubation means anything that drives the system away from its wild-type state \parencite{Visser:2003yp}},sort={4}, parent={robustness}}
\newglossaryentry{robustness5}{description={5) ``... a RBN with n attractors is robust under gene duplication and divergence if all of its n attractors are conserved after the duplication and divergence of one gene.'' \parencite{Aldana:2007da}},sort={5}, parent={robustness}}
\newglossaryentry{run}{name={Run},text={run},description={``\ldots when an apparatus has been set up and allowed for function under a specific set of experimental conditions.'' \parencite{Box2005} Essentially a synonym for \gls{replicate}, but without the sense of membership in a series under the same conditions}}

\newacronym{sd}{SD}{Standard Deviation ($\sigma$)}
\newglossaryentry{smiles}{name={SMILES},description={A human-readable notation for representing chemical molecules and reactions \parencite{smiles}}}
\newacronym{ssa}{SSA}{Stochastic Simulation Algorithm}

\newglossaryentry{trait}{name={trait},description={Any observable, phenotypic feature of an individual. \parencite{OUPBioMolBio:fr}}}
\newglossaryentry{tf}{name={Transcription factor},text={transcription factor},description={A protein that contains a DNA-binding region that when bound to DNA regulates the transcription of genetic information from DNA to RNA \parencite{Watson:2008fm}}}
\newglossaryentry{thorax}{name={thorax},description={The division of an animal's body lying between head and abdomen. In mammals, the thorax extends from neck to diagram but does not include the upper limbs. In insects, contains the wings and legs}}
\newglossaryentry{transcription}{name={transcription},description={The synthesis of a new strand of RNA complementary to a DNA template \parencite{Watson:2008fm}}}
\newglossaryentry{translation}{name={translation},description={The transfer of information from a strand of mRNA into a linear sequence of amino acids forming a protein \parencite{Watson:2008fm}}}

\newacronym{uuid}{UUID}{Universally unique identifier}

\newglossaryentry{variation}{name={variation},description={Actually present differences between the individuals in a population or a sample, or between the species in a clade. \parencite{Wagner:1996kc}}}
\newglossaryentry{variability}{name={variability},description={The potential or the propensity to vary. \parencite{Wagner:1996kc}}}