<<setup, include=FALSE>>=
library(knitr)
library(cowplot) # styling of plots, extension of ggplot2
theme_set(theme_cowplot(font_size=8))
library(gridExtra) # grid layouts for ggplot2
library(lattice) # needed for bwplot etc
library(english) # convert numbers to words
opts_chunk$set(fig.path='generated_figures/')
knit_hooks$set(pdfcrop = hook_pdfcrop)

load.toyworld <- function(f){
colClasses <- c("NULL", "NULL", "factor","NULL","factor","factor", "factor","factor","integer","integer","integer")
df <- read.csv(f,colClasses=colClasses)
df$Partition.Start <- factor(df$Partition.Start, levels = c("4750", "9750", "14750", "19750"))
df
}
@

\chapter{Introduction to Artificial Chemistries}\label{introduction-to-achems}

Artificial Chemistries provide an interesting testbed for investigating various evolutionary phenomena. This \namecref{introduction-to-achems} provides background material to aid in understanding how \glspl{achem} are almost uniquely suited to the exploration of our research questions.

Fundamentally, \glspl{achem} provide a tunable evolutionary system, capable of highly complex behaviour, built around familiar metaphors (real-world chemistry, and potentially biology). A set of rules describing how atoms interact gives rise to emergent forms--molecules. At a higher level, these molecules, under the same interaction rules, also interact in patterns--reactions.

Still higher emergent levels emerge under favourable conditions. Reactions may form cycles, where a sequence of reactions eventually returns to an earlier state. Cycles in particular are interesting as many biological processes are cyclical\footnote{In the broadest sense life can be seen as an autocatalytic process where an entity catalyses the production of one or more descendant entities.}. Replication, resulting in an exact copy of an entity, is a macro-example of a cycle; metabolism is another. Building on the apparent correspondence between higher emergent levels in \gls{achem} evolution and biology, we believe like others (\eg \cite{Steel2013}) that cycles, of some form, are a necessary building-block for more complicated structures in \glspl{achem}.

Artificial chemistries are regularly employed in three application areas: real-world chemistry simulators; tools for the exploration of artificial life, and models to test various hypotheses of the origin of life. Chemistry emulators and origins-of-life tools aim for fidelity with real-world chemistry, unlike most artificial life models. Real-world fidelity requires either the use of a library of predefined reactions, which conflicts with the goal of unlimited extension, or a chemically plausible method of constructing reactions from first principles. Because of the complexities of real-world chemistry this later method appears to be quite difficult, and the goal of a realistic, computationally practical, artificial chemistry remains open. 

However, the more limited objective of a less realistic, but still unbounded chemistry, has been achieved (see \cref{classification-of-artificial-chemistries} for examples.) One advantage of semi-realistic \glspl{achem}, as pointed out by \textcite[5]{Funes2001}, is that it is easier to evaluate solutions in a domain close to the real world as opposed to a purely symbolic or abstract domain such as a lambda-calculus, or a programmatic environment like Tierra. When contemplating difficult problems such as complexity our intuition can be helpful, but only in situations close enough to our normal experience for it to be relevant. \Glspl{achem} are a particular type of model, familiar from real world chemistry, for the simulation of reaction-based systems. 

A mathematical treatment of \glspl{achem} can be found in \textcite{Benko2009}. \Textcite{Dittrich:2001zr} and \textcite{Suzuki2008a} provide excellent reviews of the field, while an influential taxonomy is given in \textcite{Dittrich:2001zr}, described in \cref{classification-of-artificial-chemistries}. Every \gls{achem} defines a set of constitutive elements (molecules), a set of transformation rules (reactions) and a mechanism for choosing and ordering the sequence of reactions (a reactor algorithm) \parencite{Dittrich:2001zr}. These components and taxonomies are described in more detail in the next \namecref{classification-of-artificial-chemistries}.

\section{Classification of Artificial Chemistries}\label{classification-of-artificial-chemistries}

A straightforward scheme for classifying \glspl{achem} is given in \textcite{Faulconbridge2011}, based on the relationship between a molecule's representation and its properties:

\begin{itemize}
	\item \emph{Symbolic}. Symbols/molecules have no inherent meaning, so no ``implicit reactions'', only preprogrammed ones.	
	\item \emph{Structured}. One or more atoms arranged in a structure (such as a string, tree, or graph) with bonds within atoms to form molecules. This results in unlimited capability, but is computationally expensive.	
	\item \emph{Sub-symbolic}. Emergence of properties from lower-levels (\eg real chemistry, also neural networks). The symbols (for example, atoms) have an internal structure which determines the macro-properties of the chemistry.
\end{itemize}

Another system was presented by \textcite[p.132]{Nellis2012} with \glspl{achem} categorised according to a quite different set of three factors:
\begin{itemize}
	\item \emph{World}. Physical elements such as molecules and atoms.
	\item \emph{Chemistry}. The interactions between the elements.
	\item \emph{Constraints}. Restrictions on how the world and chemistry interact. Nellis gives as an example a binding model that describes which interactions are possible between elements.
\end{itemize}

However, there is an earlier, and more influential, categorisation by \textcite{Dittrich:2001zr} which also categorises \glspl{achem} against three factors: in this case, \textless{}\emph{S},\emph{R},\emph{A}\textgreater{} where \emph{S} describes the form of the component molecules in the \gls{achem}, \emph{R} the rules for the interactions between the molecules, or reaction rules, and \emph{A} the mechanism used to select molecules for reactions. 

This taxonomy is commonly used in the literature (\eg in \cite{Lenaerts2009,Gardiner2007}). In more detail, the three elements are:

\subsection{Set of Molecules (\textless{}\emph{S}\textgreater{})}

This element describes how molecules are represented in the \gls{achem} (\eg labelled graphs \parencite{Faulconbridge2011} or binary strings \parencite{Banzhaf94}), and how any implicit properties of molecules may be derived, such as bond energy calculations by Extended Huckel Theory (EHT) (\eg textcite{Benko2003}.)

\subsection{Reaction rules (\textless{}\emph{R}\textgreater{})}

Reaction rules in an artificial chemistry describe how reactants are transformed to reaction products. Reactions may be pre-defined (often the case when simulating real-world chemistry where exact pathways are important) or dynamically determined using molecular properties. In \glspl{achem} enforcing conservation of energy, elements are neither created or destroyed so reactions can be represented solely by rearrangements or bond changes. \Textcite{Tominaga2007} showed for a particular artificial chemistry, that it is computationally universal with only uni-molecular and bimolecular reactions; the reaction rules in most \glspl{achem} therefore only describe these forms of reaction.

\subsubsection{Constructive chemistries}\label{constructive-chemistries}

%\quote{A distinguishing feature of chemistry is that the changes of molecules upon interaction are not limited to quantitative physical properties such as free energy, density, or concentrations, since molecular interactions do not only produce more of what is already there--rather, novel molecules can be generated.}{\parencite{Benko2009}}

The action of the reaction rules are \textit{constructive} \parencite{Fontana1994} if new components may be generated through the action of other components--a form of emergence, explicitly linked to the production of novelties: \quote{construction: to understand how the organizations upon which the process of natural selection is based arise, and to understand how mutation can give rise to organizational, that is: phenotypic, novelty.}{\parencite{Fontana1994}}

A strongly constructive system is one which maintains closure\footnote{Or as stated by \textcite[217]{Fontana1994},``A strongly constructive system that contains agent \emph{A} must cope with the network of its implications. But, then, it also must cope with the implications of the implications. And so on.''}, and in which there is self-consistency and some form of logical structure. Both implicit laws and implicit molecule definitions are required for constructive chemistries:

\begin{itemize}
	\item Implicit reaction laws are molecular structure-based, while explicit laws are independent of molecular structure.
	\item Implicit molecule definitions provide a description for a molecule's construction, while explicit molecule definitions are taken from an fixed set of symbols.
\end{itemize}

Strongly constructive chemistries provide a mechanism for exploration in the \gls{ea} sense--new products can be generated, and new products can participate in reactions. A weakly constructive approach might be to pre-specify all possible reactions; the strongly constructive approach is to generate reactions ``on-the-fly'' from the structures of the interacting molecules. 

Pre-specification is well suited to simulating real chemistry as it allows properties of reactions observed in chemical experiments to be attached to their simulated equivalents. However, it does not allow for arbitrary reactions, and it requires reaction properties to be predetermined--rather difficult for novel or artificial reactions. Clearly, although constructive, this is not strongly constructive (in the sense of \textcite{Fontana1994,Dittrich:2001zr}) as it is not open-ended. However, \textcite{Hartenfeller2011} suggests that it is still useful for applications such as drug discovery.

An artificial chemistry with the ability to create reactions ``on-the-fly'' given a set of possible reactants may discover more than one possible reaction pathway between the same reactants and products. The method used to choose one reaction pathway from the alternatives is an important component of the artificial chemistry, and the mechanism may be tuned or tailored to privilege or preferentially chose particular types of pathways independent to other factors such as temperature or concentration. In Chemistry the choice of reaction pathway is fundamentally linked to those other properties and cannot be treated independently.

\subsection{Reactor Algorithm (\textless{}\emph{A}\textgreater{})}

The reactor algorithm provides the mechanism to select (in effect, to order) reactions, or in the words of \textcite[sect. 4.1.3]{Faulconbridge2011}, the ``...algorithm which describes the order of and intervals between reactions, starting from an initial collection of molecules...''

If the reaction rules are analogous to chemistry, the reactor algorithm is analogous to physics--the way molecules move and collide in the reaction vessel determines which reactions are possible (for example, by providing enough kinetic energy to overcome a reaction's activation energy), and the order in which the reactions occur.

\Cite{Faulconbridge2011} identifies three basic types of \emph{mixing method} or reactor algorithm:
\begin{enumerate}
	\item
	\emph{Well-mixed/aspatial}. Either discrete time (uniform probability distribution for selecting reactions) or continuous time \parencite{Gillespie1976} assuming the reactions are known in advance (which of course is not possible for a strongly constructive \gls{achem}.)
	\item
	\emph{nDimensional}. A grid with reactions in adjacent cells, or a continuous space where molecules have position and velocity. The main advantage is the ability to simulate spatial affects; the primary disadvantage is performance.
	\item
	\emph{Mixed scale}. Hierarchical spaces, such as aspatial cells within bigger grid, mostly for simulating biology (\eg \cite{Jeschke2008}). One possible advantage is the potential for parallelisation.
\end{enumerate}

\section{Applications in real-world chemistry}\label{applications-in-real-world-chemistry}

\Glspl{achem} may be applied to backward-chain from a set of desired products to identify a set of currently-available initial molecules, for example in drug discovery (\eg \cite{Hartenfeller2011}). A second use is in reaction network discovery, where the goal is to describe a closed set of reactions and reactants from some initial reactants and reactions (\eg \cite{Faulon2001}). Finally, \glspl{achem} can be used in the modelling of biological phenomenon such as enzyme function (\eg \cite{Flamm2010}.)

The primary requirement in these cases is fidelity with real-world chemistry, which requires either a library of empirically derived reaction definitions and rates, or a model capable of accurately simulating quantum-mechanical processes. The latter approach has been taken by a family of Artificial Chemistries, beginning with \textcite{Benko2003}, built on Extended H\"{u}ckel Theory with parameters taken directly from chemical experiments and later extended (for example in \cite{Benko2005}) to a general purpose model with parameters derived from theoretical chemistry. The model was used in \textcite{Hogerl2010} for the study of the behaviour and topology of chemical reaction networks, specifically Diels-Alder and Formose reaction networks, and in a series of papers (\eg \cite{Flamm2010,Ullrich2010}) for the examination of the evolution of metabolic networks in early organisms using a simple model of RNA coding for catalysts.

The core problem in drug-discovery is the selection of a set of reactions to generate a given product. The Chemical Abstract Service (CAS) database references approximately 34 million known reactions in published work; clearly it is impractical in a laboratory setting to simply apply each of these in turn to a set of initial molecules in the hope that the goal compound will emerge. \Cite{Hartenfeller2012} used simulations in reaction-SMARTS and RDKit \parencite{rdkit} to conclude that a much smaller set of 58 reactions, acting on a set of 10,000 to 50,000 ``building block'' molecules, might instead meet the requirements for \textit{de novo} drug discovery. Of the 58 nominated reactions, 29 were ring-forming; the suggestion is that as so many interesting compounds involve molecular rings, the lack of ring forming reactions was a deficiency of previous approaches. With this small reaction set, chosen for the practicality of transferral to the laboratory benchtop, the combinatorics are such that a search mechanism is needed to identify the most promising pathways from the initial molecule set to the destination compound.

Finally, \Glspl{achem} have been used as tools to explore the role of function prediction, for example in the interesting, but necessarily simplified, approach taken by \textcite{Flamm2010,Ullrich2010}. Determining the shape, and hence the function, of an enzyme from its RNA transcript is perhaps the most important problem in current molecular biology: the three-dimensional structure of a molecule--the arrangement of the elements in space--drives many of even the simplest chemical reactions. For example, acid-base reactions result from the shift of charge from one region of a molecule to another, revealing one region while shielding another from activity. Some of the most complex reactions are shape-driven: the function of many enzymes (biological catalysts) derives from their shape, and furthermore this shape is often under regulatory control. \Textcite{Flamm2010,Ullrich2010} attach a catalytic function to a molecule based on its secondary structure (shape) and then investigates the influence of these functions upon the evolution of early metabolic networks. 
	
\section{Origins of life modelling with artificial chemistries}\label{achem-and-ool}

Real-world chemical processes are also important to modelling scenarios for the origin of life or other related areas such as the formation of metabolic networks in the earliest protocells. In most cases though the specific focus is less on the bottom up model constructed from the most basic elements (although Kauffman's autocatalytic protein sets, and Kaneko's protocell toy mode are counter-examples,) and more on task-based models of processes where the particular base level is predetermined by the researcher, such as Ganti's chemoton \parencite{Ganti:2003hl}.

\Textcite{Farmer1986} describe an \gls{ode} model of polymers where bidirectional reactions connect each polymer condensate $c$ from monomer or polymer constituents $a$ and $b$, catalysed by some enzyme $e$, in the presence of water $h$. All reactions are catalysed (by some randomly chosen polymer), principally to reduce computational complexity; this is justified by the order of magnitude difference in the reaction rates between catalysed and uncatalysed reactions.  The model commences with a food set or initial population of monomers and simple polymers in a well-mixed chemostat, no pertubations, and ends when no further new polymers are generated.

Some other examples of artificial chemistries for the investigation of the origin of life include:

\begin{itemize}
\item
\emph{Lattice \gls{achem}} \parencite{Madina2003,Ono2000}. Membrane formation and cell division, assumeing five different types of particles (some hydrophilic and some hydrophobic) that together form an autocatalytic cycle similar to those observed in biological cells.
\item
\emph{SCL}. Three types of particle are employed by the Substrate-Catalyst-Link (or SCL) chemistry of \textcite{Varela:1974qd,Suzuki2008}: the eponymous Substrate ($S$), Link ($L$) and Catalyst ($C$). Cells are formed from links around a catalyst, with a single predefined reaction rule $S + S\xrightarrow{C} L$ and some straightforward constraints on movement of the particles in the matrix (for example, bonded Link particles cannot cross each other.)
\item
\emph{\Textcite{Flamm2010, Ullrich2010}}. The evolution of metabolic networks in early organisms using a simple model of RNA coding for catalysts
\item \emph{\Textcite{Hogerl2010}}. The simulation of chemical reaction networks characteristic of the transition to life, specifically the Diels-Alder and Formose reaction networks.
\item
\emph{\Textcite{Dorin:2006fk}}. Concerned, almost uniquely, with a chemical ecosystem, based on a set of atoms interacting in pre-specified ways to represent biological photosynthesis, respiration and biosynthesis (or growth). The goal is to explore the interactions in an ecosystem made up of a set of organisms pre-built to perform various defined roles.
\item
\emph{\Textcite{Gardiner2007}}. A string-based chemistry to investigate protein metabolism evolution under genetic control. Three types of molecule--protein, gene and service molecule--react in ways determined by the types of interacting molecules. The type and pattern of molecules define the type of interaction.
\item
\emph{\Textcite{Fernando:2008xy,Fernando:2007pf}}. A flow-reactor for the evolution of metabolism in lipid aggregates based on predefined molecular types and reactions.
\end{itemize}

\section{Artificial chemistries and Alife}\label{artificial-chemistries-and-alife}

Artificial Chemistries have also been used in the exploration of open-ended or creative evolution (\eg \cref{tab1}). Squirm3 \parencite{Hutton2002,Hutton2009} adopts fixed molecule types, and pre-defined reactions for replication and gene-sequence transcription, and so although capable of interesting behaviour is not capable of unlimited extension. StringMol \parencite{Hickinbotham2011}, a bacterial inspired microprogram chemistry, demonstrates a rich inheritance mechanism using string-matching as a model for molecular binding, and RBN-World \parencite{Faulconbridge2011} shows that a form of Random Boolean Network, with the addition of a bonding mechanisms to allow for composition and decomposition of RBNs, can be used to build a chemistry capable of almost limitless extension out of non-traditional components.

\begin{table}
	\scriptsize
	\begin{center}
	\caption{A sample of Artificial Chemistries for open-ended evolution.}
	\label{tab1}
	\begin{tabular}{@{}p{4cm}p{4.5cm}p{4.5cm}@{}}
		\toprule
		Chemistry                                                          & Energy Model?                                                      & Constructive?                            \\ 
		\midrule
		\cite{Ducharme2012}                                                & Yes                                                                & Yes                                      \\
		StringMol \parencite{Hickinbotham2012}                             & No, global energy only, conservation of mass as proposed extension & Yes                                      \\
		Squirm3 \parencite{Hutton2002,Lucht2012}                           & No                                                                 & No                                       \\
		RBN-World \parencite{Faulconbridge2011}                            & Unknown                                                            & Yes                                      \\
		\cite{Lenaerts2009}                                                & No                                                                 & Yes - molecular interactions             \\
		ZChem \parencite{Tominaga2004}                                     & Conservation of mass                                               & No - reactions are atomic with wildcards \\
		Substrate-Catalyst-Link (SCL) \parencite{Varela:1974qd,Suzuki2008} & No                                                                 & Unknown                                  \\
		\cite{Fernando:2008xy,Fernando:2007pf}                             & Yes, and thermodynamics govern reactions                           & No - atomic reactions                    \\
		\cite{Gardiner2007}                                                & No                                                                 & No - atomic reactions                    \\
		NAC \parencite{Suzuki2006}                                         & Unknown                                                            & Yes                                      \\						
		GGL/ToyChem \parencite{Benko2005}                                  & Mass conservation only                                             & Yes                                      \\
		Lattice \gls{achem} \parencite{Ono2000,Madina2003}        & No                                                                 & No                                       \\
		GGL/ToyChem \parencite{Benko2003}                                  & Mass conservation only                                             & No - pre-defined reactions only          \\
		\bottomrule
	\end{tabular}
	\end{center}
\end{table}

\section{Conclusion}

In this \namecref{introduction-to-achems}, we have:

\begin{itemize}
	\item Introduced the basic elements of \glspl{achem}.
	\item Established those elements within a commonly-used classification scheme \parencite{Dittrich:2001zr} which uses the three factors--\emph{S} describing the form of the component molecules, \emph{R} the rules for the interactions between the molecules, and \emph{A} the mechanism used to select molecules for reactions--to classify \glspl{achem}.
	\item Provided a reminder of the requirement for constructive reaction rules in open-ended chemical systems.
	\item Reviewed the three dominant areas for the use of \glspl{achem}: real-world chemical modeling, artificial life, and the origin of life.
	\item Suggested, by providing examples of other \glspl{achem} in the exploration of \gls{alife} and the origin of life, the suitability of an \gls{achem} for addressing our research questions.
\end{itemize}

From the brief review in \cref{achem-and-ool} of \glspl{achem} in the exploration of the origin of life it is clear that there is little consensus around the form of the preferred \gls{achem} to adopt. In the next \namecref{toyworld}, we shall move from the general description of \glspl{achem} to the introduction of a new and specific \gls{achem}, ToyWorld, that lies closer to the chemically-realistic end than to the abstract end of the \gls{achem} continuum.

\chapter{Reaction Cycles in a Molecular Artificial Chemistry}\label{toyworld}

Our chosen method for the exploration of states in variable replicators is a new molecular \gls{achem}. There is no shortage of existing \glspl{achem}; why yet another one? The primary reasons are lack of access to existing code; specific requirements of the \gls{achem} imposed by the planned series of experiments; an easy ability to modify and extend the chemistry to include specific functions, and finally, the understanding that comes from designing and building a system from the beginning.

The experimental requirements are however the primary driver. Our need is for a chemistry that is parameterised in all important aspects, so that experimental factors may be easily transformed into model parameters, and that is semi-realistic; a chemistry that shares a common language and ideally behaviour with real-world chemistry.

Other types of chemistry are likely equally suitable. But there are pragmatic reasons why a semi-realistic chemistry is appropriate. First, as previously mentioned, familiarity. Second, we can leverage existing chemical concepts--reactions, bonds, energy transformations--without needing to invent and justify a new model. Third, the most relevant previous work is from the related field of origins of life research, and this of course assumes real world chemistry. It is more likely we shall be successful in using this material to influence and inform if we adopt a closely aligned chemistry rather than something tangential. Finally, the overall argument for self-evolving evolutionary systems is postulated on emergence from low-level elements; a clear pathway exists to real-world chemistry from physics, and by analogy we can carry this correspondence through into our \gls{achem}. For example, reactants may be chosen based upon physical locations; possible reactions are driven by the energy of molecular collisions. These aspects follow naturally if we adopt a semi-realistic chemistry. 

The potential complexity of a semi-realistic chemistry is however a difficulty. No \gls{achem} can accurately model all aspects of real-world chemistry, emergent as that is from the quantum world; therefore, all semi-realistic \gls{achem} face decisions as to which elements of real-world chemistry are to be modelled and which left aside. Increasing fidelity in most cases comes at a price in performance (for example, if a quantum-physics simulator is employed to model atomic interactions) but this fidelity to real-world chemistry may in fact be unnecessary for the purposes of the \gls{achem}. And with choices comes the risk of arbitrariness--if bonds require energy to break and form, then exactly how much energy is required in each specific case? And how is this choice justified?

Therefore, our principle is that only the ``core'' elements of real-world chemistry are incorporated into our \gls{achem}--atoms, molecules, reactions. Further complexities, such as in the introduction of an energy model for the transformation between bonds and kinetic energy, are included as an optional element, controlled by a model parameter. The risk of arbitrariness is mitigated by adopting as closely as is possible real-world values, such as for bond energies. This strategy not only provides consistency, but is somewhat justifiable on the grounds of maximising where possible the correspondence with real-world chemistry.

\section{The ToyWorld artificial chemistry}

ToyWorld, our \gls{achem} for the exploration of emergent replicators, was first introduced in \textcite{Young2013}. The main elements of the model--Atoms, Molecules, Reactions, a Reaction Vessel--are recognisable from real-world chemistry, but in highly simplified forms. The most pertinent divergence is that we do not model catalytic reactions--remember that our thesis is that autocatalytic sets based on stoichiometry alone can give rise to variable replicators. 

The ToyWorld model has many degrees-of-freedom, and these must be constrained by parameter choice before the model can be used in simulation. Some values are important to our thesis, and so are considered true independent variables in our investigation. These are examined fully. The remainder however are those to which the simulation is insensitive, but still must be specified. For these we prefer real-world values rather than arbitrary artificial values. In short, where we consider something important, we investigate. Where we think it less important, we use a consistent set of pre-existing values--real-world chemistry.

The name has been chosen as both an acknowledgement of ToyChem \parencite{Benko2003}, and as a hint at the simulation's purpose: creating an artificial world for exploration. That world, the ToyWorld model, consists of:

\begin{itemize}
\item Analogues of physical elements such as atoms and molecules;
\item An overall energy model that describes the transformations that can occur between potential, kinetic and internal energy;
\item A physics model to describe how molecules interact within the reaction vessel; and
\item A chemical model that details the bond changes that can occur when two molecules collide.
\end{itemize}

All atoms, and therefore molecules and reactions, are contained within a reaction vessel. ToyWorld provides a basic energy model, where molecules have kinetic energy and bond breaking requires energy input and bond formation releases energy. The reaction vessel, which provides the strategies by which reaction reactants (or input molecules) and products (output molecules) are determined, is described in detail in the following section.

In our model, all reactions emerge solely from the properties of the reacting molecules. For each reaction between two molecules we generate a list of reaction alternatives by enumerating all possible bond additions, bond subtractions, and changes in bond type between the reactants. For example, the reactants H\textsubscript{2} and O\textsubscript{2} generate three reaction alternatives: breaking of the H-H bond, breaking of the O=O double bond, and a transformation of the O=O double bond to a single bond. The reactants H\textsuperscript{+} and OH\textsuperscript{-} give two alternative reactions: breaking of the O-H bond (giving H+H\textsuperscript{+}+O\textsuperscript{-}) and formation of a single bond between H\textsuperscript{+} and O to give H\textsubscript{2}O.

ToyWorld is custom Python code that uses third-party packages and libraries for certain functions. The most significant chunks of external functionality come from RDKit, an open-source toolkit for ChemInformatics, and PyMunk, a physics toolkit. Additionally, we use the NetworkX package for Python for network graph manipulations (mainly in the code that searches for cycles in the reaction graphs), and we use arrays from the Numpy package for Python to record the molecular populations.

RDKit provides a number of convenient functions to the ToyWorld Chemical Model:

\begin{enumerate}
	\item Format conversions between RDKit molecules and SMILES \autocite{smiles}, a standard language for molecular representation in chemistry, and conversions from the representation of a molecule into a canonical form.
	\item Reference information: for example, the mass of an atom, and the number of outer electrons for an atom.
	\item Molecular structure manipulation: iteration over the atoms or bonds in a molecule, and the addition, modification and deletions of bonds and atoms.
	\item Utility functions: combining two representations, each of one	molecule, into one representation of two molecules, and vice versa, and sanitising the representation of a molecule by checking for molecular validity.
\end{enumerate}

PyMunk is used extensively in ToyWorld for 2D physics calculations, for example:

\begin{enumerate}
	\item Calculating the future position of a molecule inside a reaction vessel based on the molecules velocity and current position.
	\item Collision detection between two (or more) molecules.
	\item Summing the forces acting on a molecule and adjusting the molecule's acceleration.
	\item Simple visualisations of the molecules within the reaction vessel as shapes on a 2D plane.
\end{enumerate}

\begin{algorithm}[ht]
	\BlankLine
	\While{generation $\leq$ generations}{
		\BlankLine
		\tcp{Selection of reactants}
		reactants $\leftarrow$ outcome of the Reactant selection strategy\;
		\BlankLine
		\tcp{Construction of reaction}
		reactions $\leftarrow$ enumerate all possible reaction options\;
		reaction $\leftarrow$ choose from $reactions$ according to the Product selection strategy\;
		\BlankLine
		\uIf{reaction is possible}{
			Update the reactor with the results of carrying out the $reaction$\;
		}
	}
	\caption{The main simulation loop in ToyWorld.}\label{alg:toyworld-main-loop}
\end{algorithm}

\section{Atoms and molecules}\label{atoms-and-molecules}

Molecules are modelled as an extension of standard RDKit \emph{Mol} objects, constructed from RDKit \emph{Atoms} connected with \emph{Bonds}. Standard Lewis dot structures built on the inherited atomic properties are used to identify possible bonds, and a formal charge model is used to record the charge changes associated with modifications to the molecular structure caused by reactions.

The lowest level component in the ToyWorld model is the atom, and atoms can be joined by bonds to form molecules. Reactions between molecules are the only mechanism to modify molecules provided by the model; a reaction is simply the addition or subtraction of a single bond between any two atoms in two molecules. 

ToyWorld provides a strongly constructive chemistry \parencite{Fontana1994} where completely new forms of molecules may be generated by reactions, and where the new molecules may in turn take part in further reactions: the chemistry emerges from the lower level atomic properties.

Some examples of molecules created by ToyWorld are shown in \cref{fig:toyworld-example-molecules}.

\begin{figure}[htb]
	\centering
	\includegraphics[width=0.45\linewidth]{figures/[H][C]N(O[N])OOC([H])(O[H])ON([O])[N][N]N(O[O])OOO[H].png}
	\includegraphics[width=0.45\linewidth]{figures/[H]OON(O[N+]([O-])[N+]([O])[O-])N([H])O[O].png}
	\includegraphics[width=0.45\linewidth]{figures/[H]OOON([H])ON([H])O[H].png}
	\includegraphics[width=0.45\linewidth]{figures/[O]C(=O)OOOON=[C]N([O])[O].png}
	\caption[Example molecules generated by ToyWorld.]{Example molecules generated by ToyWorld, rendered as 3-D objects from the original SMILES notation by MolView (\url{http://molview.org/}) - [H][C]N(O[N])OOC([H])(O[H])ON([O])[N][N]N(O[O])OOO[H] (top left), [H]OON(O[N+]([O-])[N+]([O])[O-])N([H])O[O] (top right), [H]OOON([H])ON([H])O[H] (bottom left) and [O]C(=O)OOOON=[C]N([O])[O] (bottom right).}\label{fig:toyworld-example-molecules}
\end{figure}

\section{Mass/Energy Model and Energy Transformations in ToyWorld}\label{energy-model}

Energy within ToyWorld is modeled in three forms--kinetic, internal and potential.

\begin{itemize}
\item Kinetic energy is the energy associated with the motion of a molecule, and equals $\frac{1}{2}mv^2$ where $m$ is the mass of a molecule and $v$ its velocity.
\item Internal energy is an abstraction of vibrational energy or heat energy within a molecule. Internal energy, while incidentally consistent with real-world chemical models, is actually included in ToyWorld as a major mechanism to introduce stochasticity in reactions. Without internal energy, all excess energy following a reaction must be allocated to kinetic energy (following energy conservation), and so there is an iso-map between reaction and product kinetic energies. With internal energy, we can divert an arbitrary proportion into internal energy and therefore we have a stochastic multi-map.
\item Potential energy is the energy associated with the bonds between atoms in a molecule. Creating a bond reduces the potential energy of a molecule; breaking a bond increases it. The potential energy of a molecule is the sum of the potential energy of each bond in the molecule. The state without bonds is defined as having zero potential energy, and so molecules with bonds have negative potential energy. The specific value of each bond is absolutely determined by the atomic number of the two atoms at either end of the bond, and the type of bond itself--single, double or triple. ToyWorld provides a table of relative bond values in \cref{default_bond_energies}, taken from real-world chemistry \footnote{Original source: Average Bond Dissociation Enthalpies \url{http://www.cem.msu.edu/~reusch/OrgPage/bndenrgy.htm}}. Unspecified bonds are given the average energy of specified bonds of the same bond type. Examples of molecules with associated potential energies are given in \cref{example_potential_energies}.
\end{itemize}

As the model enforces conservation of mass, reactions can be represented solely by bond changes. This follows the approach taken in graph-based chemistries such as GGL/ToyChem \parencite{Benko2003,Benko2005} where reactions are modelled as a series of changes to graph edges, or bonds, only.

\begin{table}
	\scriptsize
	\begin{center}
	\caption{Example potential energies calculated by the default Chemistry module using simplified bond energies.}\label{example_potential_energies}
	\begin{tabular}{@{}p{3cm}p{5cm}p{2.5cm}@{}}
		\toprule
		Molecule   & SMILES                   & Potential Energy \\
		\midrule
		H$_2$O     & [H]O[H]                  & -222.0           \\
		H$_2$      & [H][H]                   & -104.2           \\
		O$_2$      & O=O                      & -119.0           \\
		N$_2$O$_4$ & [O-][N+](=O)[N+]([O-])=O & -434.4           \\
		NO$_2$     & N(=O)[O]                 & -198.0           \\
		\bottomrule
	\end{tabular}
	\end{center}
\end{table}%

\begin{table}
	\scriptsize
	\begin{center}
	\caption[Bond energies in ToyWorld.]{Relative energies required to break and/or form a bond. Creating a bond between Atom 1 and Atom 2 releases molecular potential energy in the form of kinetic energy while breaking a bond does the opposite. The values in this table are those used in the experiments, and are those adopted by default in ToyWorld, but may easily be changed.}\label{default_bond_energies}
	\begin{tabular}{@{}lp{2.5cm}p{2.5cm}p{4cm}@{}}
		\toprule
		Bond Type & Atom 1 & Atom 2 & Energy of both break and formation (in simulation units)\\
		\midrule
		Single    & H      & H      & 104.2                                                \\
		Single    & C      & C      & 83                                                   \\
		Single    & N      & N      & 38.4                                                 \\
		Single    & O      & O      & 35                                                   \\
		Single    & H      & C      & 99                                                   \\
		Single    & H      & N      & 93                                                   \\
		Single    & H      & O      & 111                                                  \\
		Single    & C      & N      & 73                                                   \\
		Single    & C      & O      & 85.5                                                 \\
		Single    & N      & O      & 55                                                   \\
		Double    & C      & O      & 185                                                  \\
		Double    & C      & C      & 146                                                  \\
		Double    & N      & N      & 149                                                  \\
		Double    & O      & O      & 119                                                  \\
		Double    & C      & N      & 147                                                  \\
		Double    & N      & O      & 143                                                  \\
		Triple    & C      & O      & 258                                                  \\
		Triple    & C      & C      & 200                                                  \\
		Triple    & N      & N      & 226                                                  \\
		Triple    & C      & N      & 213                                                  \\
		Quadruple & C      & C      & 200                                                  \\
		\bottomrule
	\end{tabular}
	\end{center}
\end{table}

\subsection{Energy transformations}\label{energy-transformations}

The only energy transformations that occur in ToyWorld are those that
occur during reactions. The energy at the beginning of the reaction is
fully bound in the potential, internal and kinetic energies of the
reactants. At the moment of collision, that portion of the kinetic
energy due to the collision (kinetic energies of reactants less kinetic
energy of centre of mass) is transformed into internal energy
in the combined reactants. If a bond forms in this phase
the freed potential energy is added into this pool of internal energy;
any bond that breaks transforms internal energy into increased potential
energy. Post-collision, a portion of the pool of remaining internal
energy is transformed back into kinetic energy. The division may be all
to kinetic energy, or all to internal, or anywhere in between; as a result, we can model any collision type from fully elastic to fully inelastic collisions.

Reactions are modelled as head-on elastic collisions between two
reactants with changes to kinetic energy equalling the increase or
decrease in molecular potential energy associated with the creation,
destruction or change of order of bonds. Creation of a bond results in a
reduction of molecular potential energy and an increase to kinetic
energy; destruction results in the reverse. A change in bond type is
modelled as the sum of a bond creation and of a bond destruction. Total
energy in the system is always constant, and equal to the sum of the
initial kinetic energy of all molecules plus the sum of their potential
energies.

\begin{figure}[t]
	\begin{center}
		\vspace{10pt}
		\begin{tikzpicture}[->,>=stealth',shorten >=1pt,auto,node distance=3cm, thick,main node/.style={circle,fill=blue!20,draw, font=\sffamily\Large\bfseries,minimum size=15mm}]
		\node[main node] (PE) at (3,3) {PE};
		\node[main node] (IE) at (0,0) {IE};
		\node[main node] (KE) at (6,0) {KE};
		\path[every node/.style={font=\sffamily\small,fill=white,inner sep=1pt}]
		    (PE) edge [<->] (IE)
		    (IE) edge [<->] (KE)
		    (KE) edge [<->] (PE);
		\end{tikzpicture}
		\caption{Energy transformations in ToyWorld.}
		\label{fig:energy_transformations}
	\end{center}
\end{figure}

Reactions conserve energy, and hence the total energy in the reaction
vessel should be constant. However, energy can be explicitly added to
and removed from the system from outside. Either case is modelled as a
uniform change in all molecular internal energies (heat and vibration).
This changes the size of the pool from which the kinetic energies of the
product molecules are determined after a collision or reaction. Adding
energy to the system increases each molecule's internal energy, which
increases the size of the pool of merged internal and kinetic energies
on collision or reaction, which increases the likely kinetic energies of
the product molecules. Removing energy from the system has of course the
opposite affect, down to the point where effectively all motion and
hence all reaction activity ceases.

A reaction may be seen as two stages in sequence: first, the choice of reactants from a population of possible reactant molecules (the Reactant selection strategy, denoted here by $S_\mathrm{Reactant}$), and second, the determination of products given that set of reactants (Product selection strategy, denoted $S_\mathrm{Product}$). The total energy (that is, potential + kinetic + internal) of the system is maintained, as is the total momentum of the molecules.

\section{Reactant selection strategies: selecting reactants for a reaction}\label{reactant-selection-strategies}

\Textcite{Faulconbridge2011} describes two generic strategies for the selection of reactants--spatial and aspatial--where the primary difference is whether molecular position is a factor in reactant selection. It is possible to further generalise this scheme by considering other differentiating factors. Analogous with real-world chemistry, a cumulative scheme presents itself starting with the pure aspatial, or uniform probability strategy, and then proceeding through a spatial strategy, based on molecular kinetics, to strategies built on kinetics plus intra-molecular and external forces such as electromagnetism. 

These strategies are based on a sequence of increasing derivatives of position or location in the reaction vessel; from no position (uniform selection), through fixed position (uninteresting as we cannot have a sequence of reactions without motion) to the first derivative (velocity or kinetic selection) and finally to the second derivative (acceleration, or force selection.) 

Accordingly we adopt this novel and more detailed classification in the strategy descriptions below.

\subsection{Uniform selection}\label{uniform-selection}

In a uniform selection strategy ($S_\mathrm{Reactant} = \mathrm{Uniform}$), reactants are chosen at random with equal (uniform) probability from the population: no property of a molecule has an effect on the selection. Conceptually we have a well-stirred reaction container with no intra-molecular forces.

\subsection{Kinetic selection.}\label{kinetic-selection.}

By contrast, in a kinetic selection strategy ($S_\mathrm{Reactant} = \mathrm{Kinetic}$) molecules have spatial position (and implicitly, velocity) within some assumed reaction vessel, and selection is determined by molecular position--molecules which are spatially co-located (that is, in collision) form a reactant set. Molecules move at constant velocity until they collide with something else (either another molecule or possibly a boundary of an explicit reaction vessel) and then either react, or bounce.  Currently in our work we assume that all molecules have a fixed and common size and shape (circular in two-dimensions), irrespective of molecular formula. The algorithm for this strategy is given in \cref{alg:kinetic-selection}. At every step the molecules in the reaction vessel move $\delta t$ units along their velocity vectors, bouncing with no loss of energy off the reaction vessel sidewalls. All molecules that collide during this step are added to the tail of the $ReactantList$.

\begin{algorithm}
\Def{KineticReactantSelection(population)}{
	\uIf{|population| < 2}{
		\Return $\emptyset$
	}
	
	i$\leftarrow$0\;
	\While{|ReactantList| = 0}{
		i $\leftarrow$ i + 1\;
		Advance the position of all molecules by $\Delta$t\;
		\tcp{Check that $\Delta$t is neither too small nor too large}
		\uIf{i > 30 and (|ReactantList| = 0 or |ReactantList| > 10)}{
			Adjust $\Delta$t\;
			i$\leftarrow$0\;
		}
	}
	\tcp{Construct reactant portion of reaction}
	\While{|ReactantList| > 0}{
		Reactants$\leftarrow$pop first pair of molecules from ReactantList\;
		InitialKE$\leftarrow \sum{ke{i}}, \forall i \in$ Reactants\;
		ReactionEnergy$\leftarrow$InitialKE - KE of the Centre of Mass of the Reactants\;
		\Return Reaction(Reactants, energy=ReactionEnergy)\;
	}
	\Return $\emptyset$
}
\caption{\emph{KineticReactantSelection.}. Reactant selection strategy where colliding molecules are returned as reactants.}\label{alg:kinetic-selection}
\end{algorithm}

\subsection{Intra-molecular selection and external force selection}\label{intra-molecular-selection-and-external-force-selection}

For completeness, although not considered in the experiments that follow, more complicated forms, where molecular velocities are not constant, can be generated by the introduction of some combination of intra-molecular forces (such as electromagnetism) or external forces (such as gravity or heat.)

\section{Product selection strategies: determining the products of a reaction}\label{product-selection-strategies}

In the ToyWorld chemical model, all reactions arise solely from the properties of the reacting molecules: it therefore defines a \emph{strongly constructive} chemistry in the sense defined earlier. As ToyWorld enforces conservation of mass, reactions can be represented solely by bond changes. This is closely related conceptually to graph-based chemistries such as GGL/ToyChem \parencite{Benko2005,Benko2003}, RBN-World \parencite{Faulconbridge2011} or NAC \parencite{Suzuki2006} where reactions are modelled as a series of changes to graph edges. However, in ToyWorld, the graph is implicit rather than explicit as it is in a graph-based chemistry.

For each interaction between two molecules we generate a list of reaction alternatives by enumerating all possible single bond additions, bond subtractions, and changes in bond type between the reactants. Each alternative is the result of a single one of these changes. For example, the reactants H\textsubscript{2} and O\textsubscript{2} generate three reaction alternatives: breaking of the H-H bond, breaking of the O=O double bond, and a transformation of the O=O double bond to a single bond. The reactants H\textsuperscript{+} and OH\textsuperscript{-} give two alternative reactions: breaking of the O-H bond (giving H+H\textsuperscript{+}+O\textsuperscript{-}) and formation of a single bond between H\textsuperscript{+} and O to give H\textsubscript{2}O. 

We restrict the options to those that can be generated by a single change to the bond structure of the reactants. This does though mean that long-chain construction by real-world polymerization is not possible in ToyWorld as the modification of a double bond to a single bond (\eg C=C to CC) in each monomer must occur simultanously with the formation of the replacement CC bond between the two monomers. Long-chain molecules can be constructed from simpler components however in the equivalent to strict polymerization; therefore although this limitation prevents ToyWorld from being used in real-world domains, it is not significant in our domain.

Each reaction alternative can be completely described by the pair of the products of the reaction (that result from the single bond addition, subtraction or change) and the associated change in overall potential energy, which of course will be the same as the potential energy change of the single bond alteration.

Creation of a bond results in a reduction of molecular potential energy,
while bond destruction results in an increase. A change in bond type is
equivalent to a creation and then a destruction. The magnitude of the
change in potential energy, measured in arbitrary energy units, is taken
from a table of bond energies for each combination of atoms and bond
type (\cref{default_bond_energies}. The standard table is based upon
a simplification of real-world chemical bond energies. For example, the
creation of a H-H bond releases 104.2 units; the breaking of a C=O
double bond takes 185 energy units. \cref{example_potential_energies}
shows the potential energy for a sample of molecules.

How should we choose between alternative sets of possible products for
the same reactants? Various product strategies appear plausible: the
random choice of an alternative; the most complex alternative; least
complex; rarest; most common, and so on, but each strategy requires
effort to develop and evaluate.

\subsection{Least Energy Strategy}\label{least-energy-strategy}

When following a Least Energy strategy ($S_\mathrm{Product} = \mathrm{LeastEnergy}$) we select a reaction by choosing with uniform probability from a distribution of reaction alternatives weighted by the total of the energy changes associated with the bond changes. This biases selection towards the Least Energy alternative; the strength of the bias is determined by the degree of the weighting. \cref{fig1} shows an example of the shift in products that occurs as a result of this weighting as the overall quantity of energy in the system is changed.

Specifically, let $E_i$ denote the energy required for the bond change in the reaction option $i$. If $E_i > 0$ the reaction is exothermic, or releases energy; otherwise it is endothermic, requiring energy to proceed.

We calculate a weighted value, $e_i$, based on the combination of $E_i$ and the available energy for the reaction, $E_{avail} > 0$, as follows:

\begin{displaymath}
	e_i=
	\begin{cases}
		\lvert E_i\rvert, & (1) \text{  if $E_i < 0$;}         \\
		0,                & (2) \text{  if $E_{avail} < E_i$;} \\
		E_{avail}--E_i,   & (3) \text{  otherwise.}            
	\end{cases}
\end{displaymath}\label{weighting-calculation}

Then, for reaction option $i$ of $n$ options, $p_i = e_i / \sum\limits_{i=1}^n e_i$, where $p_i$ is the probability of option $i$ being selected. A number is chosen from the uniform distribution $\mathcal{U}[0,1]$ and the selected reaction is found from the inverse of the CDF given by $p_i\text{ for all }i$ by searching for the reaction at that point in the CDF. This method has the property that the probability of a reaction being selected is proportional to its weight.

As a result, for exothermic reactions, highly exothermic reactions are preferred to slightly exothermic ones. For endothermic reactions, the available energy must exceed the energy required by the reaction, and the reaction is preferred according to the degree of the surplus. Note that an option where the energy required exceeds that available ($E_{avail} < E_i$) will have $p_i = 0$, and hence can not be selected. This behaviour is shown in \cref{fig:reaction_selection_weights}.

\begin{figure}[t]
	\begin{center}
		\begin{tikzpicture}
			[enode/.style={circle,draw=gray!50,fill=white,radius=0.5cm}]
			\pgftext[at=\pgfpoint{2cm}{0cm},left,base]{\pgfimage[width=0.5\linewidth]{figures/reaction_selection_weights}};
			\node[enode] at (5.5,6) {1};
			\node[enode] at (7,4) {3};
			\node[enode] at (8,3.2) {2};
		\end{tikzpicture}
		\caption[Weighting calculation for reaction selection under a Least Energy Strategy.]{Weighting calculation for reaction selection under a Least Energy Strategy. Numerical labels refer to cases in \cref{weighting-calculation}.}\label{fig:reaction_selection_weights}
	\end{center}
\end{figure}

\subsection{Uniform Strategy}

We also consider a strategy with minimal bias: a Uniform selection strategy ($S_\mathrm{Product} = \mathrm{Uniform}$), where every alternative product set has equal probability of selection.

\subsection{Energy transformations during a reaction}\label{energy-transformations-during-a-reaction}

Molecules have kinetic energy; when they collide the form of the
interaction follows from the energy transformations between kinetic and
internal and potential energy that are preferred under the chemical
model; and finally, the trajectory taken by the resulting products of
the interaction is given by their final post-collision kinetic energy.

First, we determine the kinetic energy of the centre of mass of the
reactants. The available energy to drive the reaction is the total
kinetic energy of the reactants plus the internal energy of the
reactants less the kinetic energy of the centre of mass.

Consider the case where two particles of equal mass but opposite
velocity collide. The KE of the centre of mass will be zero (as it is
motionless) and the energy liberated by the collision will be the sum of
the kinetic and internal energies of the particles.

At the other extreme, consider two equal mass particles, travelling with
the same velocities. Intuitively, it is obvious that the only energy
released by their infinitesimally gentle collision is from their
internal energies, and this is confirmed by the calculation where the
kinetic energy of the centre of mass will equal the combined kinetic
energies of the reactants, leaving only the combined internal energies.

The available energy for bond modifications is calculated as the sum of
the kinetic energies of the reactants less the energy of their centre of
mass, plus any energies internal to the reactants. The final kinetic
energies of the products equals the sum of the initial kinetic energies
of reactants less the change in molecular potential energy from bond
changes and the change in internal energies. Total energy in the system
is always constant, and equal to the sum of the initial kinetic energy
of all molecules plus the sum of their potential energies and internal
energies.

The Reactor Algorithm selects one of the reaction alternatives by
choosing from a distribution of reaction alternatives weighted by
associated energy changes (as discussed in
\cref{product-selection-strategies}.)

\subsection{Setting product velocities and internal energies}\label{setting-product-velocities-and-internal-energies}

The final step in the reaction mechanism is to determine the velocities
and internal energies of the reaction products following the reaction.
Although the method is standard physics, there are two complications:
the number of products may or may not be the same as the number of
reactants, and the pre-reaction energy and post-reaction energy vary as
the reaction itself either consumes or liberates energy.

The only constraints are that velocities of the product molecules must
conserve momentum and total energy within the reacting system. We
recognise that within the frame of reference of the centre of momentum
of the reacting system, the vector sum of the momentum of the products
must equal zero. Therefore one possible solution to the product
velocities is to arrange their vector momentums according to simple
geometry: for two products, we arrange their momentums in a line (line
\ref{alg:2products} in \cref{alg:post_collision_adjustments}); for three
products, an equilateral triangle (line \ref{alg:3products}), and by
extension, for four products, a square and so on.

\begin{algorithm}
	$\text{kinetic energy of CoM}\leftarrow \frac{1}{2}(\text{sum of reactant masses})(\text{velocity of CoM})^2$\;
	\BlankLine
	$\text{collision energy}\leftarrow \text{sum of reactant kinetic energies}+\text{sum of reactant internal energies}-\text{kinetic energy of CoM}$\;
	\BlankLine
	\For(\tcp*[f]{Transform into CoM frame}){$i\leftarrow 1$ \KwTo $\text{number of Products}$}{\label{alg:lab_to_CoM_frame}
		$v^\prime_i\leftarrow v_i-\text{CoM velocity}$\;
	}
	\BlankLine
	\uIf(\tcp*[f]{Conservation of momentum implies all excess energy must go into internal energy}){Number of products = 1}{\label{alg:allocate_energies}
		$\text{Internal energy of Products}\leftarrow \text{collision energy}$\;
	}
	\uElse{
		$\text{ke}\leftarrow \text{random}([0,1])$\;
		$\text{Internal energy of Products}\leftarrow \text{collision energy}-\text{ke}$\;
	}
	\BlankLine
	\Switch(\tcp*[f]{Find a set of momentum vectors that sum to zero...}){Number of products}{
		\uCase(\tcp*[f]{One product}){1}{
			$mv^\prime\leftarrow (0,0,0)$\;
		}
		\uCase(\tcp*[f]{Two products}){2}{\label{alg:2products}
			$mv\leftarrow 2\text{ke}\prod\limits_{i=1}^n \text{mass}_i/\sum\limits_{i=1}^n \text{mass}_i$\;
			$mv^\prime_1\leftarrow (v^\prime_{i\theta}+\frac{\pi}{2},v^\prime_{i\rho}+\frac{\pi}{2},mv)$\;
			$mv^\prime_2\leftarrow (v^\prime_{i\theta}+\frac{3\pi}{2},v^\prime_{i\rho}+\frac{3\pi}{2},mv)$\;
		}
		\uCase(\tcp*[f]{Three products}){3}{\label{alg:3products}
			$mv\leftarrow 2\text{ke}\prod\limits_{i=1}^n \text{mass}_i/\sum\limits_{i=1}^n \text{mass}_i$\;
			$mv^\prime_1\leftarrow (v^\prime_{i\theta}+\frac{\pi}{3},0,mv)$\;
			$mv^\prime_2\leftarrow (v^\prime_{i\theta}-\frac{\pi}{3},0,mv)$\;
			$mv^\prime_3\leftarrow (v^\prime_{i\theta}-\pi,0,mv)$\;
		}
	}
	\BlankLine
	\For(\tcp*[f]{Convert momentums to velocities...}){$i\leftarrow 1$ \KwTo $\text{number of products}$}{
		$v^\prime_i\leftarrow mv^\prime_i / \text{mass}_i$\;
	}
	\BlankLine
	\For(\tcp*[f]{Transform back to standard frame}){$i\leftarrow 1$ \KwTo $\text{number of products}$}{\label{alg:CoM_to_lab_frame}
		$v_i\leftarrow v^\prime_i+\text{CoM velocity}$\;
	}
	\caption{Algorithm to set post-collision velocities and internal energies.}\label{alg:post_collision_adjustments}
\end{algorithm}

Following a standard method, we first transfer the molecules from the frame of reference of the reaction vessel into the centre of mass (CoM) reference frame by subtracting the velocity of the CoM from each particle (line \ref{alg:lab_to_CoM_frame}). Correspondingly, we also adjust the energy of the collision by subtracting the KE of the CoM. We recognise that in the CoM frame the vector sum of the momentums will be zero; working in this frame reduces the number of vectors we must sum by one (the momentum of the CoM itself.)

We then choose, from a uniform distribution, the proportion of total available energy to allocate to the product kinetic energies and assign the remainder to internal energy (line \ref{alg:allocate_energies}.) From the kinetic energy allocation we can determine the total scalar momentum of the products using $KE = 0.5 x velocity x momentum$, and arrange the vector momentums according to the geometry described earlier.

Finally, we convert from the CoM reference frame to the initial frame by adding back the velocity of the CoM (line \ref{alg:CoM_to_lab_frame}).  This method satisfies our requirements of conservation of momentum and energy for arbitrary numbers of reactants and products while being computationally straightforward. The limitation is that product vectors are arranged in regular and consistent, although reasonably realistic, configurations. An improvement would be to perturb the geometry of the vectors in the CoM frame to remove the regularity.

The outcomes for each reaction alternative from an example collision are shown in \cref{fig:collision_diagrams}.

\tdplotsetmaincoords{60}{110}

\begin{figure}
	\centering
	\subcaptionbox{O=C=O + C$\rightarrow$ [O] + C + [C]=O}[0.5\textwidth]{%
		\begin{tikzpicture}[scale=3,tdplot_main_coords]
			\coordinate (O) at (0,0,0);
			\draw[thick,->] (0,0,0) -- (1,0,0) node[anchor=north east]{$x$};
			\draw[thick,->] (0,0,0) -- (0,1,0) node[anchor=north west]{$y$};
			\draw[thick,->] (0,0,0) -- (0,0,1) node[anchor=south]{$z$};
			\node at (1,1,0) {};
			\tdplotsetcoord{CM0}{0.545078574376}{-35.9326309999}{-145.082834065};\tdplotsetcoord{CM1}{-0.545078574376}{-35.9326309999}{-145.082834065};
			\draw[color=black,densely dotted,-stealth] (CM0)-- (CM1);\tdplotsetcoord{I0}{-0.741455902351}{53.4742770534}{-140.637846457};
			\tdplotsetcoord{I1}{-1.00269224405}{-269.549896905}{-273.397556374};
			\draw[color=red,-stealth] (I0) node[font=\tiny,color=black]{O=C=O}-- (O);
			\draw[color=red,-stealth] (I1) node[font=\tiny,color=black]{[H]C([H])([H])[H]}-- (O);
			\draw[dotted, color=black] (O) -- (I0xy); \draw[dotted, color=black] (I0) -- (I0xy);
			\draw[dotted, color=black] (O) -- (I1xy); \draw[dotted, color=black] (I1) -- (I1xy);
			\tdplotsetcoord{O0}{0.791487199406}{-24.6589855625}{-70.1992043245};
			\tdplotsetcoord{O1}{0.790684881135}{-24.6841653158}{-219.830383842};
			\tdplotsetcoord{O2}{0.342756745318}{-57.7497471497}{-145.082834065};
			\draw[color=blue,-stealth] (O) -- (O0) node[font=\tiny,color=black]{[O]};
			\draw[color=blue,-stealth] (O) -- (O1) node[font=\tiny,color=black]{[H]C([H])([H])[H]};
			\draw[color=blue,-stealth] (O) -- (O2) node[font=\tiny,color=black]{[C]=O};
			\draw[dotted, color=black] (O) -- (O0xy); \draw[dotted, color=black] (O0) -- (O0xy);
			\draw[dotted, color=black] (O) -- (O1xy); \draw[dotted, color=black] (O1) -- (O1xy);
			\draw[dotted, color=black] (O) -- (O2xy); \draw[dotted, color=black] (O2) -- (O2xy);
		\end{tikzpicture}}%
	\subcaptionbox{O=C=O + C$\rightarrow$ C + [O][C]=O}[0.5\textwidth]{%
		\begin{tikzpicture}[scale=3,tdplot_main_coords]
			\coordinate (O) at (0,0,0);
			\draw[thick,->] (0,0,0) -- (1,0,0) node[anchor=north east]{$x$};
			\draw[thick,->] (0,0,0) -- (0,1,0) node[anchor=north west]{$y$};
			\draw[thick,->] (0,0,0) -- (0,0,1) node[anchor=south]{$z$};
			\node at (1,1,0) {};
			\tdplotsetcoord{CM0}{0.545078574376}{-35.9326309999}{-145.082834065};\tdplotsetcoord{CM1}{-0.545078574376}{-35.9326309999}{-145.082834065};
			\draw[color=black,densely dotted,-stealth] (CM0)-- (CM1);\tdplotsetcoord{I0}{-0.741455902351}{53.4742770534}{-140.637846457};
			\tdplotsetcoord{I1}{-1.00269224405}{-269.549896905}{-273.397556374};
			\draw[color=red,-stealth] (I0) node[font=\tiny,color=black]{O=C=O}-- (O);
			\draw[color=red,-stealth] (I1) node[font=\tiny,color=black]{[H]C([H])([H])[H]}-- (O);
			\draw[dotted, color=black] (O) -- (I0xy); \draw[dotted, color=black] (I0) -- (I0xy);
			\draw[dotted, color=black] (O) -- (I1xy); \draw[dotted, color=black] (I1) -- (I1xy);
			\tdplotsetcoord{O0}{0.964870972861}{170.402519717}{-86.502709589};
			\tdplotsetcoord{O1}{0.690770768471}{-121.920061352}{-167.114195829};
			\draw[color=blue,-stealth] (O) -- (O0) node[font=\tiny,color=black]{[H]C([H])([H])[H]};
			\draw[color=blue,-stealth] (O) -- (O1) node[font=\tiny,color=black]{[O][C]=O};
			\draw[dotted, color=black] (O) -- (O0xy); \draw[dotted, color=black] (O0) -- (O0xy);
			\draw[dotted, color=black] (O) -- (O1xy); \draw[dotted, color=black] (O1) -- (O1xy);
		\end{tikzpicture}}%
%\end{figure}\begin{figure}

\centering
\subcaptionbox{O=C=O + C$\rightarrow$ [O] + C + [C]=O}[0.5\textwidth][l]{%
	\begin{tikzpicture}[scale=3,tdplot_main_coords]
		\coordinate (O) at (0,0,0);
		\draw[thick,->] (0,0,0) -- (1,0,0) node[anchor=north east]{$x$};
		\draw[thick,->] (0,0,0) -- (0,1,0) node[anchor=north west]{$y$};
		\draw[thick,->] (0,0,0) -- (0,0,1) node[anchor=south]{$z$};
		\node at (1,1,0) {};
		\tdplotsetcoord{CM0}{0.545078574376}{-35.9326309999}{-145.082834065};\tdplotsetcoord{CM1}{-0.545078574376}{-35.9326309999}{-145.082834065};
		\draw[color=black,densely dotted,-stealth] (CM0)-- (CM1);\tdplotsetcoord{I0}{-0.741455902351}{53.4742770534}{-140.637846457};
		\tdplotsetcoord{I1}{-1.00269224405}{-269.549896905}{-273.397556374};
		\draw[color=red,-stealth] (I0) node[font=\tiny,color=black]{O=C=O}-- (O);
		\draw[color=red,-stealth] (I1) node[font=\tiny,color=black]{[H]C([H])([H])[H]}-- (O);
		\draw[dotted, color=black] (O) -- (I0xy); \draw[dotted, color=black] (I0) -- (I0xy);
		\draw[dotted, color=black] (O) -- (I1xy); \draw[dotted, color=black] (I1) -- (I1xy);
		\tdplotsetcoord{O0}{0.95444795957}{-20.4286793523}{-48.3296764651};
		\tdplotsetcoord{O1}{0.953114678315}{-20.4573799628}{-241.693389093};
		\tdplotsetcoord{O2}{0.239763118095}{-84.1834478971}{-145.082834065};
		\draw[color=blue,-stealth] (O) -- (O0) node[font=\tiny,color=black]{[O]};
		\draw[color=blue,-stealth] (O) -- (O1) node[font=\tiny,color=black]{[H]C([H])([H])[H]};
		\draw[color=blue,-stealth] (O) -- (O2) node[font=\tiny,color=black]{[C]=O};
		\draw[dotted, color=black] (O) -- (O0xy); \draw[dotted, color=black] (O0) -- (O0xy);
		\draw[dotted, color=black] (O) -- (O1xy); \draw[dotted, color=black] (O1) -- (O1xy);
		\draw[dotted, color=black] (O) -- (O2xy); \draw[dotted, color=black] (O2) -- (O2xy);
	\end{tikzpicture}}%
\subcaptionbox{O=C=O + C$\rightarrow$ C + [O][C]=O}[0.5\textwidth][r]{%
	\begin{tikzpicture}[scale=3,tdplot_main_coords]
		\coordinate (O) at (0,0,0);
		\draw[thick,->] (0,0,0) -- (1,0,0) node[anchor=north east]{$x$};
		\draw[thick,->] (0,0,0) -- (0,1,0) node[anchor=north west]{$y$};
		\draw[thick,->] (0,0,0) -- (0,0,1) node[anchor=south]{$z$};
		\node at (1,1,0) {};
		\tdplotsetcoord{CM0}{0.545078574376}{-35.9326309999}{-145.082834065};\tdplotsetcoord{CM1}{-0.545078574376}{-35.9326309999}{-145.082834065};
		\draw[color=black,densely dotted,-stealth] (CM0)-- (CM1);\tdplotsetcoord{I0}{-0.741455902351}{53.4742770534}{-140.637846457};
		\tdplotsetcoord{I1}{-1.00269224405}{-269.549896905}{-273.397556374};
		\draw[color=red,-stealth] (I0) node[font=\tiny,color=black]{O=C=O}-- (O);
		\draw[color=red,-stealth] (I1) node[font=\tiny,color=black]{[H]C([H])([H])[H]}-- (O);
		\draw[dotted, color=black] (O) -- (I0xy); \draw[dotted, color=black] (I0) -- (I0xy);
		\draw[dotted, color=black] (O) -- (I1xy); \draw[dotted, color=black] (I1) -- (I1xy);
		\tdplotsetcoord{O0}{0.683938755498}{116.950434333}{-109.96313961};
		\tdplotsetcoord{O1}{0.613112203007}{-91.4698581679}{-158.027937122};
		\draw[color=blue,-stealth] (O) -- (O0) node[font=\tiny,color=black]{[H]C([H])([H])[H]};
		\draw[color=blue,-stealth] (O) -- (O1) node[font=\tiny,color=black]{[O][C]=O};
		\draw[dotted, color=black] (O) -- (O0xy); \draw[dotted, color=black] (O0) -- (O0xy);
		\draw[dotted, color=black] (O) -- (O1xy); \draw[dotted, color=black] (O1) -- (O1xy);
	\end{tikzpicture}}%
%\end{figure}\begin{figure}

\centering
\subcaptionbox{O=C=O + C$\rightarrow$ [H] + [CH3] + O=C=O}[0.5\textwidth]{%
	\begin{tikzpicture}[scale=3,tdplot_main_coords]
		\coordinate (O) at (0,0,0);
		\draw[thick,->] (0,0,0) -- (1,0,0) node[anchor=north east]{$x$};
		\draw[thick,->] (0,0,0) -- (0,1,0) node[anchor=north west]{$y$};
		\draw[thick,->] (0,0,0) -- (0,0,1) node[anchor=south]{$z$};
		\node at (1,1,0) {};
		\tdplotsetcoord{CM0}{0.13448625105}{-35.9326309999}{-145.082834065};\tdplotsetcoord{CM1}{-0.13448625105}{-35.9326309999}{-145.082834065};
		\draw[color=black,densely dotted,-stealth] (CM0)-- (CM1);\tdplotsetcoord{I0}{-0.182938074093}{53.4742770534}{-140.637846457};
		\tdplotsetcoord{I1}{-0.247392444315}{-269.549896905}{-273.397556374};
		\draw[color=red,-stealth] (I0) node[font=\tiny,color=black]{O=C=O}-- (O);
		\draw[color=red,-stealth] (I1) node[font=\tiny,color=black]{[H]C([H])([H])[H]}-- (O);
		\draw[dotted, color=black] (O) -- (I0xy); \draw[dotted, color=black] (I0) -- (I0xy);
		\draw[dotted, color=black] (O) -- (I1xy); \draw[dotted, color=black] (I1) -- (I1xy);
		\tdplotsetcoord{O0}{1.0085637103}{-4.76020744144}{22.9882733079};
		\tdplotsetcoord{O1}{0.174049248064}{-27.6898552083}{-202.938127977};
		\tdplotsetcoord{O2}{0.113557100128}{-42.6716880361}{-145.082834065};
		\draw[color=blue,-stealth] (O) -- (O0) node[font=\tiny,color=black]{[H]};
		\draw[color=blue,-stealth] (O) -- (O1) node[font=\tiny,color=black]{[H][C]([H])[H]};
		\draw[color=blue,-stealth] (O) -- (O2) node[font=\tiny,color=black]{O=C=O};
		\draw[dotted, color=black] (O) -- (O0xy); \draw[dotted, color=black] (O0) -- (O0xy);
		\draw[dotted, color=black] (O) -- (O1xy); \draw[dotted, color=black] (O1) -- (O1xy);
		\draw[dotted, color=black] (O) -- (O2xy); \draw[dotted, color=black] (O2) -- (O2xy);
	\end{tikzpicture}}%
\subcaptionbox{O=C=O + C$\rightarrow$ [H] + [CH3] + O=C=O}[0.5\textwidth]{%
	\begin{tikzpicture}[scale=3,tdplot_main_coords]
		\coordinate (O) at (0,0,0);
		\draw[thick,->] (0,0,0) -- (1,0,0) node[anchor=north east]{$x$};
		\draw[thick,->] (0,0,0) -- (0,1,0) node[anchor=north west]{$y$};
		\draw[thick,->] (0,0,0) -- (0,0,1) node[anchor=south]{$z$};
		\node at (1,1,0) {};
		\tdplotsetcoord{CM0}{0.212817921117}{-35.9326309999}{-145.082834065};\tdplotsetcoord{CM1}{-0.212817921117}{-35.9326309999}{-145.082834065};
		\draw[color=black,densely dotted,-stealth] (CM0)-- (CM1);\tdplotsetcoord{I0}{-0.289490563666}{53.4742770534}{-140.637846457};
		\tdplotsetcoord{I1}{-0.391486455219}{-269.549896905}{-273.397556374};
		\draw[color=red,-stealth] (I0) node[font=\tiny,color=black]{O=C=O}-- (O);
		\draw[color=red,-stealth] (I1) node[font=\tiny,color=black]{[H]C([H])([H])[H]}-- (O);
		\draw[dotted, color=black] (O) -- (I0xy); \draw[dotted, color=black] (I0) -- (I0xy);
		\draw[dotted, color=black] (O) -- (I1xy); \draw[dotted, color=black] (I1) -- (I1xy);
		\tdplotsetcoord{O0}{1.00268330145}{-7.57832855688}{10.7762512874};
		\tdplotsetcoord{O1}{0.247189031344}{-30.8825003788}{-183.173163231};
		\tdplotsetcoord{O2}{0.193079615497}{-39.6636579632}{-145.082834065};
		\draw[color=blue,-stealth] (O) -- (O0) node[font=\tiny,color=black]{[H]};
		\draw[color=blue,-stealth] (O) -- (O1) node[font=\tiny,color=black]{[H][C]([H])[H]};
		\draw[color=blue,-stealth] (O) -- (O2) node[font=\tiny,color=black]{O=C=O};
		\draw[dotted, color=black] (O) -- (O0xy); \draw[dotted, color=black] (O0) -- (O0xy);
		\draw[dotted, color=black] (O) -- (O1xy); \draw[dotted, color=black] (O1) -- (O1xy);
		\draw[dotted, color=black] (O) -- (O2xy); \draw[dotted, color=black] (O2) -- (O2xy);
	\end{tikzpicture}}%
\caption[Reaction alternatives generated by ToyWorld from a collision between O=C=O and C molecules.]{Reaction alternatives generated by ToyWorld from a collision between O=C=O and C molecules. Each alternative generates a unique set of products, with corresponding velocities calculated by \cref{alg:post_collision_adjustments} so as to preserve energy and momentum.}
\label{fig:collision_diagrams}
\end{figure}

\section{Validation of the energy model}\label{model-validation}

The energy-model presented in this \namecref{toyworld} leads to some particular expectations for the behaviour of ToyWorld:

\begin{enumerate}
\item Given two reactants, changing the reaction energy should result in different sets of reaction products.
\item Molecular quantities should reach equilibrium--that is, the set of interacting molecules is constant, with fluctuations expected in quantities. We expect molecular concentrations to stabilise at non-extreme values (equilibrium rather than driven to an extreme) after some transition period from the initial conditions.
\item The equilibrium point should depend on the energy of the system. Our energy model preferentially forms bonds at low energies, and breaks bonds at high. We expect the average length of molecules in the artificial chemistry to be greater at low energies than at high energies.
\end{enumerate}

These predictions were tested by two experiments: first, we examined the reaction products produced at a range of reaction energies for four sets of reactants. Second, for a given set of reactants, we ran the simulation for 10,000 iterations at four successive initial average kinetic energy levels--0, 67, 133, and 200 units per molecule. The experiment was run first with a reactant set containing 100xN\textsubscript{2}O\textsubscript{4} and 100x2NO\textsubscript{2}, and then with a reactant set of 100xH\textsubscript{2}, 100xO\textsubscript{2} and 100xH\textsubscript{2}O.

\begin{figure}
	\begin{center}
		\caption[Molecular quantities over time for initial population of N\textsubscript{2}O\textsubscript{4} and 2NO\textsubscript{2}.]{Molecular quantities over time for initial population of N\textsubscript{2}O\textsubscript{4} and 2NO\textsubscript{2} with initial average KE ranging from 0 to 200 units (only molecules with significant quantities are labelled; remainder appear as light-grey lines). Molecules represented in SMILES notation.}
		\includegraphics[width=\linewidth]{figures/results-test-N2O4a}s
		\label{fig1}
	\end{center}
\end{figure}

\subsection{Results and discussion}\label{results-and-discussion}

Beyond the initial transition period, both reactant sets showed results
essentially consistent with equilibrium. Population variability was high
in both cases, but more so for the N\textsubscript{2}O\textsubscript{4}
and 2NO\textsubscript{2} reactant set (see \cref{fig1}) where some molecules
never reached a relatively constant population level (gradient of a
best-fit population line remained significantly non-zero.) The
fluctuations in the quantities of the other molecules are expected
according to our criteria, and result from the inherent variability in
reaction selection which causes the quantities to oscillate around a
norm.

With both reactant sets the model produced a significant number of
molecules which would be considered unstable in real-world chemistry
(such as O\textsuperscript{-} and O.) This is likely an artefact of the method
we use to generate reaction options, where a bond-break plus
bond-formation reaction--moving through an intermediate unstable
ion--occurs in our model as two separate reactions. As all molecules
currently react with equal likelihood, significant time can elapse
before the intermediate product reacts to form a stable product.

Both reactant sets showed clear differences in population composition
between the four initial kinetic energy levels. In the
H\textsubscript{2}, O\textsubscript{2} and H\textsubscript{2}O reactant set, no reactions
occurred at the zero energy level. This is expected from our energy
model as only bond-formations are possible without free kinetic energy.
With reactants of H\textsubscript{2}, O\textsubscript{2} and H\textsubscript{2}O no bond formations are possible,
confirmed by examining the bond options returned by the model for the
six possible combinations of initial reactants. By contrast, the
reactant set N\textsubscript{2}O\textsubscript{4} and
2NO\textsubscript{2} at energy zero contains one possible bond formation
reaction (in SMILES, [O]N=O.[O]N=O to O=N[O][O]N=O)
which can proceed without free kinetic energy. This then releases a
product which can also react, and so on, thus explaining the different
results between the reaction sets.

The energy and reaction models produce results consistent with our predictions for the system's behaviour (with the exception of achieving equilibrium with the N\textsubscript{2}O\textsubscript{4} and 2NO\textsubscript{2} reactant set). An aspatial approach does however come with restrictions. Most obviously, as there is no concept of proximity, there can be no boundaries or membranes or even basic distinctions between \emph{inside} and \emph{outside}. This is critical in biology but it is unclear at this point if this is equally important in non-biological systems. An experimental comparison between the aspatial and spatial approaches in the experiments described in the next section will help to clarify this.

\section{Sensitivity of the model to reactant and product strategies}\label{reactant-and-product-strategies}

The ToyWorld model contains a number of parameters that might conceivably affect its response; as the results of any experiment, and hence the conclusions drawn, are causally related to the experimental parameters used, it's important to characterize the influence of parameters on the results. In this section we explore the following questions:

\vspace{0.3cm}
\begin{minipage}[l]{0.95\textwidth}
	\begin{enumerate}[label=RQ2.\arabic*:]
		\item Is there a quantitative difference between different reactant and product selection strategies?
		\item Is there a combination of reactant and product selection strategies that leads to increased emergence as measured by cycles?
		\item Is emergence significantly affected by the values of other parameters of an \gls{achem}, such as initial kinetic energy or bond energies?
	\end{enumerate}
\end{minipage}
\vspace{0.3cm}

To the best of our knowledge, this is the first time that reaction and product selection strategies in \glspl{achem} have been experimentally compared.
Instead, the general approach of previous work, where there has been a quantitative evaluation, has been to propose a particular strategy, build, and evaluate against the initial goals, rather than against alternatives.

Our two primary factors, or independent variables, are $S_\mathrm{Reactant}$ and $S_\mathrm{Product}$. We also introduce two secondary factors, overall reaction vessel energy ($E_\mathrm{Vessel}$) and bond energy ($E_\mathrm{Bonds}$), to assess the sensitivity of the simulation to other parameters.

For simplicity of analysis, all of our factors are two-level, meaning they take one of two possible levels, or values, in each run. The parameter values chosen for each level of $E_\mathrm{Vessel}$ and $E_\mathrm{Bonds}$ were chosen as representative from a set of alternatives used in initial exploratory experiments; in each case they allowed the simulation to run for an extended period without running out of possible reactions (from lack of energy for example.)

\begin{table}
	\scriptsize
	\begin{center}
	\caption{Factors, or independent variables.}\label{tbl:factors}
	\begin{tabular}{p{1.4cm}p{2.2cm}p{4.4cm}p{5.5cm}}
		\toprule
		Factor                & +1 value                          & -1 value                                                                                        & Description                                                    \\
		\midrule
		\noalign{\smallskip}
		$S_\mathrm{Reactant}$ & Kinetic                           & Uniform                                                                                         & See Section \cref{reactant-selection-strategies}               \\
		$S_\mathrm{Product}$  & LeastEnergy                       & Uniform                                                                                         & See Section \cref{product-selection-strategies}                \\
		$E_\mathrm{Vessel}$   & 300                               & 100                                                                                             & Initial kinetic energy of each molecule in the reaction vessel \\
		$E_\mathrm{Bonds}$    & Single=50, Double=100, Triple=200 & Simplified real-world chemistry. Average values for Single=77.7, Double=148.2, and Triple=224.3 & Energy required to break a bond of the given type              \\
		\bottomrule
	\end{tabular}
	\end{center}
\end{table}

We concentrate on three related response, or dependent, variables--Number of cycles, Length of longest cycle, and Count of most common cycle. All three are derived from a reconstruction of the network of reactions that occur during each experiment run, where every edge represents a specific reaction connecting a particular set of reactants with a particular set of products. Note that the nodes in the constructed network capture specific molecules, rather than molecular types or species that share the same chemical formula (as would be more usual in the construction of a Reaction Network for real-world chemistry.)

We exclude all unique cycles, and all cycles with three or fewer elements (for example, where a molecule loses, then regains, an atom repeatedly). Unique cycles by nature are unlikely to be representative; very short cycles on the other hand are so common as to dominate other more interesting cycles in any analysis.

\section{Experiment design}
The experiments follow a full factorial design over four factors ($S_\mathrm{Reactant}$, $S_\mathrm{Product}$, $E_\mathrm{Vessel}$ and $E_\mathrm{Bonds}$), each at two levels, run in a randomised order, with three (3) replicates of each combination of factors executed in sequence before beginning the next combination. The first replicate of each combination starts with a predefined random seed incremented by one for each successive replicate of the same combination. The factor levels used are given in \cref{tbl:factors}.

Each replicate used the same initial population of 800 molecules, made up of 100 molecules each of [H][H], O=O, [O-][N+](=O)[N+]([O-])=O, and N(=O)[O] and 200 molecules each of O and O=C=O (all represented in SMILES \parencite{smiles}.) This initial population is somewhat arbitrary, although reasonable; given that ToyWorld is a strongly constructive chemistry, we would expect that any differences between initial populations would reduce as the simulation proceeds.

\section{Results}

The overall sequence of experiments to run was defined by an experiment design. The experiments were run in the order they were defined, with all replicates of an experiment being completed before moving to the replicates of the next experiment in the sequence. All replicates of an experiment were run with the same set of parameters with the exception of a random seed which varied between replicates in a predictable way. The reactions in the replicate were grouped into fixed-size blocks and run in sequence with results saved incrementally between each block.

All replicates completed a set of 20,000 reactions; given the initial population size of 800 molecules, and from the summary of results below, we believe that this captures a representative set of reactions. This also simplifies the analysis as we can assume a balanced set of treatments in the statistical sense (that is, the sample sizes for all treatments are equal).

\begin{table}[htbp]
	\scriptsize
	\begin{center}
	\caption{Summary of results.}\label{tbl:results}
	\begin{tabular}{lrrr}
		\toprule
		Statistic    & Number of cycles & Length of longest cycle & Count of most common cycle \\
		\hline\noalign{\smallskip}
		\multicolumn{4}{c}{Reactions 4750 to 5000}\\
		\hline\noalign{\smallskip}
		Min.         & 0.00             & 0.00                    & 0.00                       \\
		1st Quartile & 0.00             & 0.00                    & 0.00                       \\
		Median       & 1.50             & 3.50                    & 2.50                       \\
		Mean         & 219.06           & 5.04                    & 215.80                     \\
		3rd Quartile & 91.25            & 7.75                    & 96.00                      \\
		Max.         & 5704.00          & 20.00                   & 2728.00                    \\
		\hline\noalign{\smallskip}
		\multicolumn{4}{c}{Reactions 9750 to 10000}\\
		\hline\noalign{\smallskip}
		Min.         & 0.00             & 0.00                    & 0.00                       \\
		1st Quartile & 0.00             & 0.00                    & 0.00                       \\
		Median       & 6.00             & 4.00                    & 6.00                       \\
		Mean         & 62.10            & 4.65                    & 169.21                     \\
		3rd Quartile & 68.75            & 8.25                    & 27.75                      \\
		Max.         & 526.00           & 13.00                   & 6684.00                    \\
		\hline\noalign{\smallskip}
		\multicolumn{4}{c}{Reactions 14750 to 15000}\\
		\hline\noalign{\smallskip}
		Min.         & 0.00             & 0.00                    & 0.00                       \\
		1st Quartile & 1.00             & 3.00                    & 2.00                       \\
		Median       & 5.00             & 4.50                    & 5.00                       \\
		Mean         & 27.17            & 4.79                    & 42.27                      \\
		3rd Quartile & 34.50            & 7.00                    & 16.25                      \\
		Max.         & 237.00           & 12.00                   & 862.00                     \\
		\hline\noalign{\smallskip}
		\multicolumn{4}{c}{Reactions 19750 to 20000}\\
		\hline\noalign{\smallskip}
		Min.         & 0.00             & 0.00                    & 0.00                       \\
		1st Quartile & 0.00             & 0.00                    & 0.00                       \\
		Median       & 3.50             & 4.00                    & 4.00                       \\
		Mean         & 20.04            & 3.90                    & 14.62                      \\
		3rd Quartile & 20.25            & 6.00                    & 13.25                      \\
		Max.         & 199.00           & 12.00                   & 216.00                     \\
		\bottomrule
	\end{tabular}
	\end{center}
\end{table}%

A view of the results is given in \cref{tbl:results}: reaction networks built from the full dataset of 20,000 reactions can be too large for easy analysis. Instead, we choose to partition the reaction data into four equally spaced blocks of 250 reactions each and analyse each block independently.

<<cyclesbypartition, pdfcrop=TRUE, echo=FALSE, cache=TRUE, warning=FALSE, fig.show='hold', fig.scap=NA, fig.cap='Cycles by reaction partition (starting reaction number for each partition along x-axis, y-axis scales vary)'>>=
df<-load.toyworld("results/EvaluatorActualCycles.out")

ap <- ggplot(df) + geom_boxplot(aes(y=Number.of.cycles, x=Partition.Start)) + labs(title='Cycle numbers by Partition', x='Start of partition', y='Number of cycles') + scale_y_continuous(trans="log10", breaks=c(0,1,2,3,4,5,6,7,8,9,10,100,1000))
bp <- ggplot(df) + geom_boxplot(aes(y=Length.of.longest.cycle, x=Partition.Start)) + labs(title='Cycle length by Partition', x='Start of partition', y='Length of longest cycle')
cp <- ggplot(df) + geom_boxplot(aes(y=Count.of.most.common.cycle, x=Partition.Start)) + labs(title='Cycle count by Partition', x='Start of partition', y='Count of most common cycle') + scale_y_continuous(trans="log10", breaks=c(0,1,2,3,4,5,6,7,8,9,10,100,1000,5000))
grid.arrange(ap,bp,cp,nrow=1,ncol=3)
@
\label{fig:partitions}

%\begin{figure}[h]
%\centering
%	\includegraphics[width=0.95\linewidth]{figures/partitions}
%	\caption{Cycles by reaction partition (starting reaction number for each partition along x-axis)}
%	\label{fig:partitions}
%\end{figure}

% No R code...
\begin{figure}[h]
	\centering
	\includegraphics[width=0.45\linewidth]{figures/PlotMolecularDiversity-strategies-12-0}
	\includegraphics[width=0.45\linewidth]{figures/PlotMolecularDiversity-strategies-16-1}
	\caption{Diversity for Replicates 12-0 and 16-1.}\label{fig:diversity}
\end{figure}

\section{Analysis and Discussion}
\Cref{fig:cyclesbypartition} suggests that the first partition, representing the vessel a quarter of the way into its lifespan, is quantitatively different from the other three partitions, with a significantly greater range for all three response variables. Intuitively this corresponds with an initial period where the diversity in the reaction vessel rapidly increases from the limited starting set of molecules, as seen in some (\eg \cref{fig:diversity}) but not necessarily all of the replicates. Diversity here is measured by (average molecular quantity)$^{-1}$. All following sections therefore exclude data from the first partition of reaction numbers from 4750 to 5000.

<<reactantstrategy, pdfcrop=TRUE, echo=FALSE, cache=TRUE, warning=FALSE, fig.show='hold', fig.scap=NA, fig.cap='Response by Reactant Strategy (y-axis scales vary)'>>=
df <- subset(df,df$Partition.Start != "4750")
ap <- ggplot(df) + geom_boxplot(aes(y=Number.of.cycles, x=Product.Strategy)) + labs(title="Log cycle numbers by Product Strategy", y="Number of cycles") + scale_x_discrete(labels=c("Uniform","Energy")) + scale_y_continuous(trans="log10", breaks=c(0,1,2,3,4,5,6,7,8,9,10,100,1000))
bp <- ggplot(df) + geom_boxplot(aes(y=Length.of.longest.cycle, x=Product.Strategy)) + labs(title="Cycle length by Product Strategy", y="Length of longest cycle") + scale_x_discrete(labels=c("Uniform","Energy"))
cp <- ggplot(df) + geom_boxplot(aes(y=Count.of.most.common.cycle, x=Product.Strategy)) + labs(title="Log cycle count by Product Strategy", y="Count of most common cycle")  + scale_x_discrete(labels=c("Uniform","Energy")) + scale_y_continuous(trans="log10", breaks=c(0,1,2,3,4,5,6,7,8,9,10,100,1000))
grid.arrange(ap,bp,cp,nrow=1,ncol=3)
@

%\begin{figure}[t]
%\centering
%\includegraphics[width=0.95\linewidth]{figures/reactant_strategy}
%\caption{Response by $S_\mathrm{Reactant}$}
%\label{fig:reactant_strategy}
%\end{figure}

<<productstrategy, pdfcrop=TRUE, echo=FALSE, cache=TRUE, warning=FALSE, fig.show='hold', fig.scap=NA, fig.cap='Response by Product Strategy (y-axis scales vary)'>>=
ap <- ggplot(df) + geom_boxplot(aes(y=Number.of.cycles, x=Reactant.Strategy)) + labs(title="Log cycle numbers by Reactant Strategy", y="Number of cycles") + scale_x_discrete(labels=c("Uniform","Kinetic")) + scale_y_continuous(trans="log10", breaks=c(0,1,2,3,4,5,6,7,8,9,10,100,1000))
bp <- ggplot(df) + geom_boxplot(aes(y=Length.of.longest.cycle, x=Reactant.Strategy)) + labs(title="Cycle length by Reactant Strategy", y="Length of longest cycle") + scale_x_discrete(labels=c("Uniform","Kinetic"))
cp <- ggplot(df) + geom_boxplot(aes(y=Count.of.most.common.cycle, x=Reactant.Strategy)) + labs(title="Log cycle count by Reactant Strategy", y="Count of most common cycle")  + scale_x_discrete(labels=c("Uniform","Kinetic")) + scale_y_continuous(trans="log10", breaks=c(0,1,2,3,4,5,6,7,8,9,10,100,1000))
grid.arrange(ap,bp,cp,nrow=1,ncol=3)
@

%\begin{figure}[t]
%\centering
%\includegraphics[width=0.95\linewidth]{figures/product_strategy}
%\caption{Response by $S_\mathrm{Product}$}
%\label{fig:product_strategy}
%\end{figure}

\subsection{RQ2.1: Is there a quantitative difference between the different reactant and product selection strategies?}
From visual inspection of \cref{fig:reactantstrategy}, there appears to be a significant difference between the Uniform and Kinetic reactant selection strategies for number and length of cycles. Kinetic reactant selection seems to result in significantly higher levels of emergent behaviour than Uniform reactant selection. Similarly, from \cref{fig:productstrategy}, there is very little apparent difference between the two product strategies, Uniform selection and Least Energy selection.

We use \gls{anova} to further examine the relationship of $S_\mathrm{Reactant}$ and $S_\mathrm{Product}$ to the response variables using a two-factor with two-levels (2x2) model with interaction effects. There is a highly significant difference (p\textless 0.001) between the Uniform and Kinetic reactant selection strategies when comparing the number of cycles (f-value=40.442) and length of cycles (f-value=361.891) (confirming the impression given by \cref{fig:reactantstrategy}), although again without difference for the count of the most common cycle. The effect of $S_\mathrm{Product}$ on cycle number and length is also significant (f-value=4.050 and 5.705 respectively, p\textless 0.05) and there is a first-order interaction between $S_\mathrm{Reactant}$ and $S_\mathrm{Product}$ for number of cycles (f-value=4.011, p\textless 0.05).

<<reactantproductcombination, pdfcrop=TRUE, echo=FALSE, cache=TRUE, warning=FALSE, fig.show='hold', fig.scap=NA, fig.cap='Effect of the combination of Reactant and Product Strategies on the number of cycles (left), and length of longest cycle (right.)'>>=
ap <- ggplot(df) + geom_boxplot(aes(y=Number.of.cycles, x=interaction(Reactant.Strategy,Product.Strategy))) + scale_y_continuous(trans="log10", limits=c(1,1000), breaks=c(0,1,2,3,4,5,6,7,8,9,10,100,1000)) + scale_x_discrete(labels=c("Uniform:Uniform","Kinetic:Uniform","Uniform:Energy","Kinetic:Energy"))+ labs(x='', y='Number of cycles')
bp <- ggplot(df) + geom_boxplot(aes(y=Length.of.longest.cycle, x=interaction(Reactant.Strategy,Product.Strategy))) + scale_y_continuous(trans="log10", limits=c(1,1000), breaks=c(0,1,2,3,4,5,6,7,8,9,10,100,1000)) + scale_x_discrete(labels=c("Uniform:Uniform","Kinetic:Uniform","Uniform:Energy","Kinetic:Energy")) + labs(x='', y='Length of longest cycle')
grid.arrange(ap,bp,nrow=1,ncol=2)
@

%\begin{figure}[t]
%\centering
%\subcaptionbox{Cycle Count by Strategy Combination %($S_\mathrm{Reactant}$:$S_\mathrm{Product}$)\label{fig:cycle_count}}[0.45\linewidth][r]{\includegraphics[width=0.45\linewidth]{figures/cycle_count}}
%\subcaptionbox{Cycle Length by Strategy Combination %($S_\mathrm{Reactant}$:$S_\mathrm{Product}$)\label{fig:cycle_length}}[0.45\linewidth][l]{\includegraphics[width=0.45\linewidth]{figures/cycle_length}}
%\subcaptionbox{Count of Most Common Cycle by Strategy Combination %($S_\mathrm{Reactant}$:$S_\mathrm{Product}$)\label{fig:cycle_common}}[0.45\linewidth][r]{\includegraphics[width=0.45\linewidth]{figures/cycle_common}}
%\caption{Effect of the combination of $S_\mathrm{Reactant}$ and $S_\mathrm{Product}$ on Response Variables}
%\end{figure}

\subsection{RQ2.2: Is there a combination of reactant and product selection strategies that leads to increased emergence as measured by cycles?}

From \cref{fig:reactantproductcombination} it is clear that there is no significant relationship between strategy and the number of occurrence of the most common cycle. However, it seems that such a relationship does exist for the number and length of cycles, with the strongest effect as a result of $S_\mathrm{Reactant}$, and a lesser effect from the choice of $S_\mathrm{Product}$.

We conclude that the greatest levels of emergence are likely to be seen with the combination of $S_\mathrm{Reactant} = \mathrm{Kinetic}$ and $S_\mathrm{Product} = \mathrm{LeastEnergy}$.

\subsection{RQ2.3: Is emergence significantly affected by the values of other parameters of an \gls{achem}, such as initial kinetic energy or bond energies?}

We construct a two-factor with two-levels (2x2) \gls{anova} model (degrees of freedom=1) with interaction effects to examine the relationship of the independent variables $E_\mathrm{Vessel}$ and $E_\mathrm{Bonds}$ to the response variables, and applied it to our dataset (summarised in \cref{tbl:results}). $E_\mathrm{Bonds}$ is significant (f-value=4.221, p\textless 0.05) to number of cycles. No other significant relationships exist.

\section{Conclusions}
In this \namecref{toyworld} we have explored the effect of four possible combinations of Reactant and Product selection strategies.

By experiment we have shown that the choice of $S_\mathrm{Reactant}$ is critical to the behaviour of this \gls{achem}; $S_\mathrm{Product}$ on the other hand appears to have a lesser effect on the emergence of cycles in our experiments. Furthermore, $S_\mathrm{Reactant} = \mathrm{Kinetic}$ is more effective for cycle emergence than $S_\mathrm{Reactant} = \mathrm{Uniform}$.

The most significant limitation of our work is that the values chosen for the high and low values of $E_\mathrm{Bonds}$ make it impossible to determine the cause of the difference observed in RQ2.3. There are two alternative explanations: first, the energy required to make or break bonds is simply different between the two factor levels; second, in the low factor level, based on real-world values, the bond make and break energies for even a single bond vary depending on the atoms involved, while in the high factor level these values are consistent for all bonds of the same degree. To distinguish between the two explanations, the average levels at each degree should be the same for each factor; this is a suggestion for a future experiment.